{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4b39c240",
   "metadata": {},
   "source": [
    "# Variant Calling Analysis\n",
    "Beased on [Circulating tumor DNA sequencing in colorectal cancer patients treated with first-line chemotherapy with anti-EGFR](https://www.nature.com/articles/s41598-021-95345-4). \n",
    "___\n",
    "\n",
    "Analysis by: Luis Aguilera. November 15, 2025\n",
    "_____"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c693c1e",
   "metadata": {},
   "source": [
    "## Project Description:\n",
    "This pipeline analyzes circulating tumor DNA sequencing data to track KRAS mutations in colorectal cancer patients during anti-EGFR treatment. \n",
    "This code aims to identify variant allele frequency changes that serve as biomarkers for monitoring treatment response and tumor evolution.\n",
    "_____"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82fc8f15",
   "metadata": {},
   "source": [
    "<img src=\"docs/image.png\" alt=\"VCA pipeline\" width=\"900\" />"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4902d374",
   "metadata": {},
   "source": [
    "### Import libraries\n",
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2e6f329f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/envs/vca_env/lib/python3.12/site-packages/pysradb/utils.py:14: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from tqdm.autonotebook import tqdm\n"
     ]
    }
   ],
   "source": [
    "# importing libraries\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "from pysradb import SRAweb \n",
    "import requests\n",
    "import pysam\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "834dded5",
   "metadata": {},
   "source": [
    "### Paths to data\n",
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "776ac9ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current working directory: /Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis\n"
     ]
    }
   ],
   "source": [
    "# current working directory\n",
    "cwd = Path.cwd()\n",
    "print(\"Current working directory:\", cwd)\n",
    "# Data directory\n",
    "data_dir = cwd / \"data\"\n",
    "data_dir.mkdir(parents=True, exist_ok=True)\n",
    "# Genome reference directory\n",
    "genome_ref_dir = data_dir / \"genome_reference\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f982784e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Utility function to run shell commands\n",
    "def run_cmd(cmd):\n",
    "    subprocess.run(cmd, check=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c3ae846",
   "metadata": {},
   "source": [
    "### Downloading the reference genome (GRCh38)\n",
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1ecb92ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reference genome already exists.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/genome_reference/Homo_sapiens.GRCh38.dna.primary_assembly.fa')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def download_reference_genome(fasta_url, genome_ref_directory):\n",
    "    \"\"\"\n",
    "    Download and decompress human reference genome.\n",
    "    Input: fasta_url (str), genome_ref_directory (Path)\n",
    "    Output: Path to decompressed FASTA file\n",
    "    \"\"\"\n",
    "    # get the name from the URL\n",
    "    gz_filename = Path(fasta_url).name\n",
    "    fa_filename = gz_filename.replace(\".gz\", \"\")\n",
    "    gz_path = genome_ref_directory / gz_filename # Path to compressed file\n",
    "    fa_path = genome_ref_directory / fa_filename # Path to decompressed FASTA file\n",
    "    if not fa_path.exists():\n",
    "        if not gz_path.exists():\n",
    "            # Download the gzipped fasta file\n",
    "            run_cmd([\"wget\", \"-O\", str(gz_path), fasta_url])\n",
    "        # Decompress the fasta file\n",
    "        run_cmd([\"gunzip\", \"-k\", str(gz_path)])\n",
    "    else:\n",
    "        print(\"Reference genome already exists.\")\n",
    "    return fa_path\n",
    "\n",
    "# Download reference genome\n",
    "human_genome_fasta_url = (\"https://ftp.ensembl.org/pub/release-109/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz\")\n",
    "fa_ref_genome_path = download_reference_genome(human_genome_fasta_url, genome_ref_dir)\n",
    "fa_ref_genome_path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10cad90c",
   "metadata": {},
   "source": [
    "### Get metadata for bioproject: PRJNA714799\n",
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "020d3962",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "run_accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "study_accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "study_title",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "experiment_accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "experiment_title",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "experiment_desc",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "organism_taxid",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "organism_name",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "library_name",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "library_strategy",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "library_source",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "library_selection",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "library_layout",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "sample_accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "sample_title",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "biosample",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "bioproject",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "instrument",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "instrument_model",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "instrument_model_desc",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "total_spots",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "total_size",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "run_total_spots",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "run_total_bases",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "run_alias",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "public_filename",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "public_size",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "public_date",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "public_md5",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "public_version",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "public_semantic_name",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "public_supertype",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "public_sratoolkit",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "aws_url",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "aws_free_egress",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "aws_access_type",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "public_url",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ncbi_url",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ncbi_free_egress",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ncbi_access_type",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "gcp_url",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "gcp_free_egress",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "gcp_access_type",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "experiment_alias",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "isolate",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "age",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "biomaterial_provider",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "sex",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "tissue",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "biosamplemodel",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ena_fastq_http",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "ena_fastq_http_1",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ena_fastq_http_2",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ena_fastq_ftp",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "ena_fastq_ftp_1",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ena_fastq_ftp_2",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "d6d0bba6-57d2-4151-bd36-c5c2d3a1550c",
       "rows": [
        [
         "0",
         "SRR13973703",
         "SRP310811",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "SRX10351697",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "9606",
         "Homo sapiens",
         "plasma_267",
         "Targeted-Capture",
         "GENOMIC",
         "Hybrid Selection",
         "PAIRED",
         "SRS8470544",
         null,
         "SAMN18318231",
         "PRJNA714799",
         "Illumina HiSeq 2500",
         "Illumina HiSeq 2500",
         "ILLUMINA",
         "6546936",
         "1218165846",
         "6546936",
         "1928631066",
         "CTC362-2.final.bam",
         "SRR13973703.lite",
         "135422449",
         "2022-09-13 13:12:48",
         "30c9f43dc571fb4ae41d2e8c5ed990eb",
         "1",
         "SRA Lite",
         "Primary ETL",
         "1",
         "s3://sra-pub-zq-6/SRR13973703/SRR13973703.lite.1",
         "s3.us-east-1",
         "aws identity",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos4/sra-pub-zq-1/SRR013/13973/SRR13973703/SRR13973703.lite.1",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos4/sra-pub-zq-1/SRR013/13973/SRR13973703/SRR13973703.lite.1",
         "worldwide",
         "anonymous",
         "gs://sra-pub-zq-104/SRR13973703/SRR13973703.lite.1",
         "gs.us-east1",
         "gcp identity",
         null,
         "blood",
         "not applicable",
         "03080, 101, Daehak-ro, Jongno-gu, Seoul, Republic of Korea",
         "not applicable",
         "plasma_267",
         "Human",
         null,
         "ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/003/SRR13973703/SRR13973703_1.fastq.gz",
         "http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/003/SRR13973703/SRR13973703_2.fastq.gz",
         null,
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/003/SRR13973703/SRR13973703_1.fastq.gz",
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/003/SRR13973703/SRR13973703_2.fastq.gz"
        ],
        [
         "1",
         "SRR13973704",
         "SRP310811",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "SRX10351696",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "9606",
         "Homo sapiens",
         "plasma_266",
         "Targeted-Capture",
         "GENOMIC",
         "Hybrid Selection",
         "PAIRED",
         "SRS8470543",
         null,
         "SAMN18318230",
         "PRJNA714799",
         "Illumina HiSeq 2500",
         "Illumina HiSeq 2500",
         "ILLUMINA",
         "8159199",
         "1364764800",
         "8159199",
         "2399054404",
         "CTC360.final.bam",
         "SRR13973704.lite",
         "148079133",
         "2022-09-13 13:12:11",
         "ae6eb4113ce33961e01cbdf59d010146",
         "1",
         "SRA Lite",
         "Primary ETL",
         "1",
         "s3://sra-pub-zq-2/SRR13973704/SRR13973704.lite.1",
         "s3.us-east-1",
         "aws identity",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos4/sra-pub-zq-1/SRR013/13973/SRR13973704/SRR13973704.lite.1",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos4/sra-pub-zq-1/SRR013/13973/SRR13973704/SRR13973704.lite.1",
         "worldwide",
         "anonymous",
         "gs://sra-pub-zq-104/SRR13973704/SRR13973704.lite.1",
         "gs.us-east1",
         "gcp identity",
         null,
         "blood",
         "not applicable",
         "03080, 101, Daehak-ro, Jongno-gu, Seoul, Republic of Korea",
         "not applicable",
         "plasma_266",
         "Human",
         null,
         "ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/004/SRR13973704/SRR13973704_1.fastq.gz",
         "http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/004/SRR13973704/SRR13973704_2.fastq.gz",
         null,
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/004/SRR13973704/SRR13973704_1.fastq.gz",
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/004/SRR13973704/SRR13973704_2.fastq.gz"
        ],
        [
         "2",
         "SRR13973705",
         "SRP310811",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "SRX10351695",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "9606",
         "Homo sapiens",
         "plasma_265",
         "Targeted-Capture",
         "GENOMIC",
         "Hybrid Selection",
         "PAIRED",
         "SRS8470542",
         null,
         "SAMN18318229",
         "PRJNA714799",
         "Illumina HiSeq 2500",
         "Illumina HiSeq 2500",
         "ILLUMINA",
         "5630116",
         "995898454",
         "5630116",
         "1664298424",
         "CTC360-2.final.bam",
         "SRR13973705.lite",
         "106879043",
         "2022-09-14 16:36:06",
         "11cf7b649a36fae2b7d143789604ec54",
         "1",
         "SRA Lite",
         "Primary ETL",
         "1",
         "s3://sra-pub-zq-6/SRR13973705/SRR13973705.lite.1",
         "s3.us-east-1",
         "aws identity",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos1/sra-pub-zq-38/SRR013/13973/SRR13973705/SRR13973705.lite.1",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos1/sra-pub-zq-38/SRR013/13973/SRR13973705/SRR13973705.lite.1",
         "worldwide",
         "anonymous",
         "gs://sra-pub-zq-104/SRR13973705/SRR13973705.lite.1",
         "gs.us-east1",
         "gcp identity",
         null,
         "blood",
         "not applicable",
         "03080, 101, Daehak-ro, Jongno-gu, Seoul, Republic of Korea",
         "not applicable",
         "plasma_265",
         "Human",
         null,
         "ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/005/SRR13973705/SRR13973705_1.fastq.gz",
         "http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/005/SRR13973705/SRR13973705_2.fastq.gz",
         null,
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/005/SRR13973705/SRR13973705_1.fastq.gz",
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/005/SRR13973705/SRR13973705_2.fastq.gz"
        ],
        [
         "3",
         "SRR13973706",
         "SRP310811",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "SRX10351694",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "9606",
         "Homo sapiens",
         "PBMC_35",
         "Targeted-Capture",
         "GENOMIC",
         "Hybrid Selection",
         "PAIRED",
         "SRS8470541",
         null,
         "SAMN18317923",
         "PRJNA714799",
         "Illumina HiSeq 2500",
         "Illumina HiSeq 2500",
         "ILLUMINA",
         "3915985",
         "458166471",
         "3915985",
         "1165468860",
         "PBMC_CTC003.final.bam",
         "SRR13973706.lite",
         "50615734",
         "2022-09-13 13:11:42",
         "852f83fb98a902ef87cac413f3e63251",
         "1",
         "SRA Lite",
         "Primary ETL",
         "1",
         "s3://sra-pub-zq-2/SRR13973706/SRR13973706.lite.1",
         "s3.us-east-1",
         "aws identity",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos4/sra-pub-zq-1/SRR013/13973/SRR13973706/SRR13973706.lite.1",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos4/sra-pub-zq-1/SRR013/13973/SRR13973706/SRR13973706.lite.1",
         "worldwide",
         "anonymous",
         "gs://sra-pub-zq-104/SRR13973706/SRR13973706.lite.1",
         "gs.us-east1",
         "gcp identity",
         null,
         "blood",
         "not applicable",
         "03080, 101, Daehak-ro, Jongno-gu, Seoul, Republic of Korea",
         "not applicable",
         "PBMC_35",
         "Human",
         null,
         "ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/006/SRR13973706/SRR13973706_1.fastq.gz",
         "http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/006/SRR13973706/SRR13973706_2.fastq.gz",
         null,
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/006/SRR13973706/SRR13973706_1.fastq.gz",
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/006/SRR13973706/SRR13973706_2.fastq.gz"
        ],
        [
         "4",
         "SRR13973707",
         "SRP310811",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "SRX10351693",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "9606",
         "Homo sapiens",
         "plasma_264",
         "Targeted-Capture",
         "GENOMIC",
         "Hybrid Selection",
         "PAIRED",
         "SRS8470540",
         null,
         "SAMN18318228",
         "PRJNA714799",
         "Illumina HiSeq 2500",
         "Illumina HiSeq 2500",
         "ILLUMINA",
         "5072064",
         "812338961",
         "5072064",
         "1498289667",
         "CTC358.final.bam",
         "SRR13973707.lite",
         "82275790",
         "2022-09-16 07:06:00",
         "739bc3d3eda894071752a0f0e812d91f",
         "1",
         "SRA Lite",
         "Primary ETL",
         "1",
         "s3://sra-pub-zq-7/SRR13973707/SRR13973707.lite.1",
         "s3.us-east-1",
         "aws identity",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos1/sra-pub-zq-38/SRR013/13973/SRR13973707/SRR13973707.lite.1",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos1/sra-pub-zq-38/SRR013/13973/SRR13973707/SRR13973707.lite.1",
         "worldwide",
         "anonymous",
         "gs://sra-pub-zq-109/SRR13973707/SRR13973707.lite.1",
         "gs.us-east1",
         "gcp identity",
         null,
         "blood",
         "not applicable",
         "03080, 101, Daehak-ro, Jongno-gu, Seoul, Republic of Korea",
         "not applicable",
         "plasma_264",
         "Human",
         null,
         "ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/007/SRR13973707/SRR13973707_1.fastq.gz",
         "http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/007/SRR13973707/SRR13973707_2.fastq.gz",
         null,
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/007/SRR13973707/SRR13973707_1.fastq.gz",
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/007/SRR13973707/SRR13973707_2.fastq.gz"
        ]
       ],
       "shape": {
        "columns": 56,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>study_title</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>experiment_desc</th>\n",
       "      <th>organism_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>library_name</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>...</th>\n",
       "      <th>biomaterial_provider</th>\n",
       "      <th>sex</th>\n",
       "      <th>tissue</th>\n",
       "      <th>biosamplemodel</th>\n",
       "      <th>ena_fastq_http</th>\n",
       "      <th>ena_fastq_http_1</th>\n",
       "      <th>ena_fastq_http_2</th>\n",
       "      <th>ena_fastq_ftp</th>\n",
       "      <th>ena_fastq_ftp_1</th>\n",
       "      <th>ena_fastq_ftp_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR13973703</td>\n",
       "      <td>SRP310811</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>SRX10351697</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>plasma_267</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>...</td>\n",
       "      <td>03080, 101, Daehak-ro, Jongno-gu, Seoul, Repub...</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>plasma_267</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/003/SRR139...</td>\n",
       "      <td>http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/003...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR13973704</td>\n",
       "      <td>SRP310811</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>SRX10351696</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>plasma_266</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>...</td>\n",
       "      <td>03080, 101, Daehak-ro, Jongno-gu, Seoul, Repub...</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>plasma_266</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/004/SRR139...</td>\n",
       "      <td>http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/004...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR13973705</td>\n",
       "      <td>SRP310811</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>SRX10351695</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>plasma_265</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>...</td>\n",
       "      <td>03080, 101, Daehak-ro, Jongno-gu, Seoul, Repub...</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>plasma_265</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/005/SRR139...</td>\n",
       "      <td>http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/005...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR13973706</td>\n",
       "      <td>SRP310811</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>SRX10351694</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>PBMC_35</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>...</td>\n",
       "      <td>03080, 101, Daehak-ro, Jongno-gu, Seoul, Repub...</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>PBMC_35</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/006/SRR139...</td>\n",
       "      <td>http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/006...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR13973707</td>\n",
       "      <td>SRP310811</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>SRX10351693</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>plasma_264</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>...</td>\n",
       "      <td>03080, 101, Daehak-ro, Jongno-gu, Seoul, Repub...</td>\n",
       "      <td>not applicable</td>\n",
       "      <td>plasma_264</td>\n",
       "      <td>Human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/007/SRR139...</td>\n",
       "      <td>http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/007...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 56 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession study_accession  \\\n",
       "0   SRR13973703       SRP310811   \n",
       "1   SRR13973704       SRP310811   \n",
       "2   SRR13973705       SRP310811   \n",
       "3   SRR13973706       SRP310811   \n",
       "4   SRR13973707       SRP310811   \n",
       "\n",
       "                                         study_title experiment_accession  \\\n",
       "0  plasma cfdna sequencing in CRC treated with 1s...          SRX10351697   \n",
       "1  plasma cfdna sequencing in CRC treated with 1s...          SRX10351696   \n",
       "2  plasma cfdna sequencing in CRC treated with 1s...          SRX10351695   \n",
       "3  plasma cfdna sequencing in CRC treated with 1s...          SRX10351694   \n",
       "4  plasma cfdna sequencing in CRC treated with 1s...          SRX10351693   \n",
       "\n",
       "                                    experiment_title  \\\n",
       "0  plasma cfdna sequencing in CRC treated with 1s...   \n",
       "1  plasma cfdna sequencing in CRC treated with 1s...   \n",
       "2  plasma cfdna sequencing in CRC treated with 1s...   \n",
       "3  plasma cfdna sequencing in CRC treated with 1s...   \n",
       "4  plasma cfdna sequencing in CRC treated with 1s...   \n",
       "\n",
       "                                     experiment_desc  organism_taxid  \\\n",
       "0  plasma cfdna sequencing in CRC treated with 1s...            9606   \n",
       "1  plasma cfdna sequencing in CRC treated with 1s...            9606   \n",
       "2  plasma cfdna sequencing in CRC treated with 1s...            9606   \n",
       "3  plasma cfdna sequencing in CRC treated with 1s...            9606   \n",
       "4  plasma cfdna sequencing in CRC treated with 1s...            9606   \n",
       "\n",
       "  organism_name library_name  library_strategy  ...  \\\n",
       "0  Homo sapiens   plasma_267  Targeted-Capture  ...   \n",
       "1  Homo sapiens   plasma_266  Targeted-Capture  ...   \n",
       "2  Homo sapiens   plasma_265  Targeted-Capture  ...   \n",
       "3  Homo sapiens      PBMC_35  Targeted-Capture  ...   \n",
       "4  Homo sapiens   plasma_264  Targeted-Capture  ...   \n",
       "\n",
       "                                biomaterial_provider             sex  \\\n",
       "0  03080, 101, Daehak-ro, Jongno-gu, Seoul, Repub...  not applicable   \n",
       "1  03080, 101, Daehak-ro, Jongno-gu, Seoul, Repub...  not applicable   \n",
       "2  03080, 101, Daehak-ro, Jongno-gu, Seoul, Repub...  not applicable   \n",
       "3  03080, 101, Daehak-ro, Jongno-gu, Seoul, Repub...  not applicable   \n",
       "4  03080, 101, Daehak-ro, Jongno-gu, Seoul, Repub...  not applicable   \n",
       "\n",
       "       tissue biosamplemodel  ena_fastq_http  \\\n",
       "0  plasma_267          Human             NaN   \n",
       "1  plasma_266          Human             NaN   \n",
       "2  plasma_265          Human             NaN   \n",
       "3     PBMC_35          Human             NaN   \n",
       "4  plasma_264          Human             NaN   \n",
       "\n",
       "                                    ena_fastq_http_1  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/003/SRR139...   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/004/SRR139...   \n",
       "2  ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/005/SRR139...   \n",
       "3  ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/006/SRR139...   \n",
       "4  ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/007/SRR139...   \n",
       "\n",
       "                                    ena_fastq_http_2 ena_fastq_ftp  \\\n",
       "0  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/003...           NaN   \n",
       "1  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/004...           NaN   \n",
       "2  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/005...           NaN   \n",
       "3  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/006...           NaN   \n",
       "4  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/007...           NaN   \n",
       "\n",
       "                                     ena_fastq_ftp_1  \\\n",
       "0  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...   \n",
       "1  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...   \n",
       "2  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...   \n",
       "3  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...   \n",
       "4  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...   \n",
       "\n",
       "                                     ena_fastq_ftp_2  \n",
       "0  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...  \n",
       "1  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...  \n",
       "2  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...  \n",
       "3  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...  \n",
       "4  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...  \n",
       "\n",
       "[5 rows x 56 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Download SRA metadata for project PRJNA714799\n",
    "metadata_file_path = data_dir.joinpath(\"SRA_metadata_PRJNA714799_full.csv\")\n",
    "if metadata_file_path.exists():\n",
    "    metadata = pd.read_csv(metadata_file_path)\n",
    "else:\n",
    "    db = SRAweb()\n",
    "    metadata = db.sra_metadata(\"PRJNA714799\", detailed=True)\n",
    "    metadata.to_csv(metadata_file_path, index=False)\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a9de885e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample_type\n",
      "plasma        271\n",
      "PBMC           76\n",
      "colorectal     73\n",
      "\n",
      "Unique patient IDs: 271\n",
      "\n",
      "isolate\n",
      "blood            347\n",
      "tissue biopsy     73\n"
     ]
    }
   ],
   "source": [
    "# Extract sample type and ID from 'tissue' column\n",
    "metadata['sample_type'] = metadata['tissue'].str.extract(r'^([A-Za-z]+)_')[0]\n",
    "metadata['sample_id'] = metadata['tissue'].str.extract(r'_(\\d+)$')[0].astype(int)\n",
    "print(f\"{metadata['sample_type'].value_counts().to_string()}\\n\")\n",
    "print(f\"Unique patient IDs: {metadata['sample_id'].nunique()}\\n\")\n",
    "print(f\"{metadata['isolate'].value_counts().to_string()}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab550a6d",
   "metadata": {},
   "source": [
    "### Extracting patient_id and timepoints from metadata\n",
    "___\n",
    "\n",
    "The SRA metadata use a consistent `run_alias` pattern:\n",
    "\n",
    "- PBMC samples: `PBMC_<ID>.final.bam`  \n",
    "  patient_id = <ID>, sample_type = PBMC, no timepoint\n",
    "\n",
    "- FFPE tumor samples: `FFPE_<ID>.final.bam`  \n",
    "  patient_id = <ID>, sample_type = Tissue, no timepoint\n",
    "\n",
    "- ctDNA samples: `<ID>.final.bam` or `<ID>-<N>.final.bam`  \n",
    "  patient_id = <ID>, sample_type = ctDNA, timepoint 0 (no suffix) or timepoint N (suffix -N)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e14d259d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "run_accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "study_accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "study_title",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "experiment_accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "experiment_title",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "experiment_desc",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "organism_taxid",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "organism_name",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "library_name",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "library_strategy",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "library_source",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "library_selection",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "library_layout",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "sample_accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "sample_title",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "biosample",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "bioproject",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "instrument",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "instrument_model",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "instrument_model_desc",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "total_spots",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "total_size",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "run_total_spots",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "run_total_bases",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "run_alias",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "public_filename",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "public_size",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "public_date",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "public_md5",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "public_version",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "public_semantic_name",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "public_supertype",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "public_sratoolkit",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "aws_url",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "aws_free_egress",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "aws_access_type",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "public_url",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ncbi_url",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ncbi_free_egress",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ncbi_access_type",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "gcp_url",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "gcp_free_egress",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "gcp_access_type",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "experiment_alias",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "isolate",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "age",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "biomaterial_provider",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "sex",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "tissue",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "biosamplemodel",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ena_fastq_http",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "ena_fastq_http_1",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ena_fastq_http_2",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ena_fastq_ftp",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "ena_fastq_ftp_1",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ena_fastq_ftp_2",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "sample_id",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "patient_id",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "timepoint",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "sample_type",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "443d201d-60d0-46ea-bd0d-3c3787384694",
       "rows": [
        [
         "0",
         "SRR13973703",
         "SRP310811",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "SRX10351697",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "9606",
         "Homo sapiens",
         "plasma_267",
         "Targeted-Capture",
         "GENOMIC",
         "Hybrid Selection",
         "PAIRED",
         "SRS8470544",
         null,
         "SAMN18318231",
         "PRJNA714799",
         "Illumina HiSeq 2500",
         "Illumina HiSeq 2500",
         "ILLUMINA",
         "6546936",
         "1218165846",
         "6546936",
         "1928631066",
         "CTC362-2.final.bam",
         "SRR13973703.lite",
         "135422449",
         "2022-09-13 13:12:48",
         "30c9f43dc571fb4ae41d2e8c5ed990eb",
         "1",
         "SRA Lite",
         "Primary ETL",
         "1",
         "s3://sra-pub-zq-6/SRR13973703/SRR13973703.lite.1",
         "s3.us-east-1",
         "aws identity",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos4/sra-pub-zq-1/SRR013/13973/SRR13973703/SRR13973703.lite.1",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos4/sra-pub-zq-1/SRR013/13973/SRR13973703/SRR13973703.lite.1",
         "worldwide",
         "anonymous",
         "gs://sra-pub-zq-104/SRR13973703/SRR13973703.lite.1",
         "gs.us-east1",
         "gcp identity",
         null,
         "blood",
         "not applicable",
         "03080, 101, Daehak-ro, Jongno-gu, Seoul, Republic of Korea",
         "not applicable",
         "plasma_267",
         "Human",
         null,
         "ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/003/SRR13973703/SRR13973703_1.fastq.gz",
         "http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/003/SRR13973703/SRR13973703_2.fastq.gz",
         null,
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/003/SRR13973703/SRR13973703_1.fastq.gz",
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/003/SRR13973703/SRR13973703_2.fastq.gz",
         "267",
         "CTC362",
         "2.0",
         "ctDNA"
        ],
        [
         "1",
         "SRR13973704",
         "SRP310811",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "SRX10351696",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "9606",
         "Homo sapiens",
         "plasma_266",
         "Targeted-Capture",
         "GENOMIC",
         "Hybrid Selection",
         "PAIRED",
         "SRS8470543",
         null,
         "SAMN18318230",
         "PRJNA714799",
         "Illumina HiSeq 2500",
         "Illumina HiSeq 2500",
         "ILLUMINA",
         "8159199",
         "1364764800",
         "8159199",
         "2399054404",
         "CTC360.final.bam",
         "SRR13973704.lite",
         "148079133",
         "2022-09-13 13:12:11",
         "ae6eb4113ce33961e01cbdf59d010146",
         "1",
         "SRA Lite",
         "Primary ETL",
         "1",
         "s3://sra-pub-zq-2/SRR13973704/SRR13973704.lite.1",
         "s3.us-east-1",
         "aws identity",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos4/sra-pub-zq-1/SRR013/13973/SRR13973704/SRR13973704.lite.1",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos4/sra-pub-zq-1/SRR013/13973/SRR13973704/SRR13973704.lite.1",
         "worldwide",
         "anonymous",
         "gs://sra-pub-zq-104/SRR13973704/SRR13973704.lite.1",
         "gs.us-east1",
         "gcp identity",
         null,
         "blood",
         "not applicable",
         "03080, 101, Daehak-ro, Jongno-gu, Seoul, Republic of Korea",
         "not applicable",
         "plasma_266",
         "Human",
         null,
         "ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/004/SRR13973704/SRR13973704_1.fastq.gz",
         "http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/004/SRR13973704/SRR13973704_2.fastq.gz",
         null,
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/004/SRR13973704/SRR13973704_1.fastq.gz",
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/004/SRR13973704/SRR13973704_2.fastq.gz",
         "266",
         "CTC360",
         "0.0",
         "ctDNA"
        ],
        [
         "2",
         "SRR13973705",
         "SRP310811",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "SRX10351695",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "9606",
         "Homo sapiens",
         "plasma_265",
         "Targeted-Capture",
         "GENOMIC",
         "Hybrid Selection",
         "PAIRED",
         "SRS8470542",
         null,
         "SAMN18318229",
         "PRJNA714799",
         "Illumina HiSeq 2500",
         "Illumina HiSeq 2500",
         "ILLUMINA",
         "5630116",
         "995898454",
         "5630116",
         "1664298424",
         "CTC360-2.final.bam",
         "SRR13973705.lite",
         "106879043",
         "2022-09-14 16:36:06",
         "11cf7b649a36fae2b7d143789604ec54",
         "1",
         "SRA Lite",
         "Primary ETL",
         "1",
         "s3://sra-pub-zq-6/SRR13973705/SRR13973705.lite.1",
         "s3.us-east-1",
         "aws identity",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos1/sra-pub-zq-38/SRR013/13973/SRR13973705/SRR13973705.lite.1",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos1/sra-pub-zq-38/SRR013/13973/SRR13973705/SRR13973705.lite.1",
         "worldwide",
         "anonymous",
         "gs://sra-pub-zq-104/SRR13973705/SRR13973705.lite.1",
         "gs.us-east1",
         "gcp identity",
         null,
         "blood",
         "not applicable",
         "03080, 101, Daehak-ro, Jongno-gu, Seoul, Republic of Korea",
         "not applicable",
         "plasma_265",
         "Human",
         null,
         "ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/005/SRR13973705/SRR13973705_1.fastq.gz",
         "http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/005/SRR13973705/SRR13973705_2.fastq.gz",
         null,
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/005/SRR13973705/SRR13973705_1.fastq.gz",
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/005/SRR13973705/SRR13973705_2.fastq.gz",
         "265",
         "CTC360",
         "2.0",
         "ctDNA"
        ],
        [
         "3",
         "SRR13973706",
         "SRP310811",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "SRX10351694",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "9606",
         "Homo sapiens",
         "PBMC_35",
         "Targeted-Capture",
         "GENOMIC",
         "Hybrid Selection",
         "PAIRED",
         "SRS8470541",
         null,
         "SAMN18317923",
         "PRJNA714799",
         "Illumina HiSeq 2500",
         "Illumina HiSeq 2500",
         "ILLUMINA",
         "3915985",
         "458166471",
         "3915985",
         "1165468860",
         "PBMC_CTC003.final.bam",
         "SRR13973706.lite",
         "50615734",
         "2022-09-13 13:11:42",
         "852f83fb98a902ef87cac413f3e63251",
         "1",
         "SRA Lite",
         "Primary ETL",
         "1",
         "s3://sra-pub-zq-2/SRR13973706/SRR13973706.lite.1",
         "s3.us-east-1",
         "aws identity",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos4/sra-pub-zq-1/SRR013/13973/SRR13973706/SRR13973706.lite.1",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos4/sra-pub-zq-1/SRR013/13973/SRR13973706/SRR13973706.lite.1",
         "worldwide",
         "anonymous",
         "gs://sra-pub-zq-104/SRR13973706/SRR13973706.lite.1",
         "gs.us-east1",
         "gcp identity",
         null,
         "blood",
         "not applicable",
         "03080, 101, Daehak-ro, Jongno-gu, Seoul, Republic of Korea",
         "not applicable",
         "PBMC_35",
         "Human",
         null,
         "ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/006/SRR13973706/SRR13973706_1.fastq.gz",
         "http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/006/SRR13973706/SRR13973706_2.fastq.gz",
         null,
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/006/SRR13973706/SRR13973706_1.fastq.gz",
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/006/SRR13973706/SRR13973706_2.fastq.gz",
         "35",
         "CTC003",
         null,
         "PBMC"
        ],
        [
         "4",
         "SRR13973707",
         "SRP310811",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "SRX10351693",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "plasma cfdna sequencing in CRC treated with 1st line chemotherapy with anti-EGFR",
         "9606",
         "Homo sapiens",
         "plasma_264",
         "Targeted-Capture",
         "GENOMIC",
         "Hybrid Selection",
         "PAIRED",
         "SRS8470540",
         null,
         "SAMN18318228",
         "PRJNA714799",
         "Illumina HiSeq 2500",
         "Illumina HiSeq 2500",
         "ILLUMINA",
         "5072064",
         "812338961",
         "5072064",
         "1498289667",
         "CTC358.final.bam",
         "SRR13973707.lite",
         "82275790",
         "2022-09-16 07:06:00",
         "739bc3d3eda894071752a0f0e812d91f",
         "1",
         "SRA Lite",
         "Primary ETL",
         "1",
         "s3://sra-pub-zq-7/SRR13973707/SRR13973707.lite.1",
         "s3.us-east-1",
         "aws identity",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos1/sra-pub-zq-38/SRR013/13973/SRR13973707/SRR13973707.lite.1",
         "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos1/sra-pub-zq-38/SRR013/13973/SRR13973707/SRR13973707.lite.1",
         "worldwide",
         "anonymous",
         "gs://sra-pub-zq-109/SRR13973707/SRR13973707.lite.1",
         "gs.us-east1",
         "gcp identity",
         null,
         "blood",
         "not applicable",
         "03080, 101, Daehak-ro, Jongno-gu, Seoul, Republic of Korea",
         "not applicable",
         "plasma_264",
         "Human",
         null,
         "ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/007/SRR13973707/SRR13973707_1.fastq.gz",
         "http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/007/SRR13973707/SRR13973707_2.fastq.gz",
         null,
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/007/SRR13973707/SRR13973707_1.fastq.gz",
         "era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/007/SRR13973707/SRR13973707_2.fastq.gz",
         "264",
         "CTC358",
         "0.0",
         "ctDNA"
        ]
       ],
       "shape": {
        "columns": 60,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>study_title</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>experiment_desc</th>\n",
       "      <th>organism_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>library_name</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>...</th>\n",
       "      <th>ena_fastq_http</th>\n",
       "      <th>ena_fastq_http_1</th>\n",
       "      <th>ena_fastq_http_2</th>\n",
       "      <th>ena_fastq_ftp</th>\n",
       "      <th>ena_fastq_ftp_1</th>\n",
       "      <th>ena_fastq_ftp_2</th>\n",
       "      <th>sample_id</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>sample_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR13973703</td>\n",
       "      <td>SRP310811</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>SRX10351697</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>plasma_267</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/003/SRR139...</td>\n",
       "      <td>http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/003...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "      <td>267</td>\n",
       "      <td>CTC362</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ctDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR13973704</td>\n",
       "      <td>SRP310811</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>SRX10351696</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>plasma_266</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/004/SRR139...</td>\n",
       "      <td>http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/004...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "      <td>266</td>\n",
       "      <td>CTC360</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ctDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR13973705</td>\n",
       "      <td>SRP310811</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>SRX10351695</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>plasma_265</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/005/SRR139...</td>\n",
       "      <td>http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/005...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "      <td>265</td>\n",
       "      <td>CTC360</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ctDNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR13973706</td>\n",
       "      <td>SRP310811</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>SRX10351694</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>PBMC_35</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/006/SRR139...</td>\n",
       "      <td>http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/006...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "      <td>35</td>\n",
       "      <td>CTC003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PBMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR13973707</td>\n",
       "      <td>SRP310811</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>SRX10351693</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>plasma cfdna sequencing in CRC treated with 1s...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>plasma_264</td>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/007/SRR139...</td>\n",
       "      <td>http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/007...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "      <td>era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...</td>\n",
       "      <td>264</td>\n",
       "      <td>CTC358</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ctDNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  run_accession study_accession  \\\n",
       "0   SRR13973703       SRP310811   \n",
       "1   SRR13973704       SRP310811   \n",
       "2   SRR13973705       SRP310811   \n",
       "3   SRR13973706       SRP310811   \n",
       "4   SRR13973707       SRP310811   \n",
       "\n",
       "                                         study_title experiment_accession  \\\n",
       "0  plasma cfdna sequencing in CRC treated with 1s...          SRX10351697   \n",
       "1  plasma cfdna sequencing in CRC treated with 1s...          SRX10351696   \n",
       "2  plasma cfdna sequencing in CRC treated with 1s...          SRX10351695   \n",
       "3  plasma cfdna sequencing in CRC treated with 1s...          SRX10351694   \n",
       "4  plasma cfdna sequencing in CRC treated with 1s...          SRX10351693   \n",
       "\n",
       "                                    experiment_title  \\\n",
       "0  plasma cfdna sequencing in CRC treated with 1s...   \n",
       "1  plasma cfdna sequencing in CRC treated with 1s...   \n",
       "2  plasma cfdna sequencing in CRC treated with 1s...   \n",
       "3  plasma cfdna sequencing in CRC treated with 1s...   \n",
       "4  plasma cfdna sequencing in CRC treated with 1s...   \n",
       "\n",
       "                                     experiment_desc  organism_taxid  \\\n",
       "0  plasma cfdna sequencing in CRC treated with 1s...            9606   \n",
       "1  plasma cfdna sequencing in CRC treated with 1s...            9606   \n",
       "2  plasma cfdna sequencing in CRC treated with 1s...            9606   \n",
       "3  plasma cfdna sequencing in CRC treated with 1s...            9606   \n",
       "4  plasma cfdna sequencing in CRC treated with 1s...            9606   \n",
       "\n",
       "  organism_name library_name  library_strategy  ... ena_fastq_http  \\\n",
       "0  Homo sapiens   plasma_267  Targeted-Capture  ...            NaN   \n",
       "1  Homo sapiens   plasma_266  Targeted-Capture  ...            NaN   \n",
       "2  Homo sapiens   plasma_265  Targeted-Capture  ...            NaN   \n",
       "3  Homo sapiens      PBMC_35  Targeted-Capture  ...            NaN   \n",
       "4  Homo sapiens   plasma_264  Targeted-Capture  ...            NaN   \n",
       "\n",
       "                                    ena_fastq_http_1  \\\n",
       "0  ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/003/SRR139...   \n",
       "1  ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/004/SRR139...   \n",
       "2  ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/005/SRR139...   \n",
       "3  ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/006/SRR139...   \n",
       "4  ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/007/SRR139...   \n",
       "\n",
       "                                    ena_fastq_http_2 ena_fastq_ftp  \\\n",
       "0  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/003...           NaN   \n",
       "1  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/004...           NaN   \n",
       "2  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/005...           NaN   \n",
       "3  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/006...           NaN   \n",
       "4  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/007...           NaN   \n",
       "\n",
       "                                     ena_fastq_ftp_1  \\\n",
       "0  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...   \n",
       "1  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...   \n",
       "2  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...   \n",
       "3  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...   \n",
       "4  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...   \n",
       "\n",
       "                                     ena_fastq_ftp_2 sample_id patient_id  \\\n",
       "0  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...       267     CTC362   \n",
       "1  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...       266     CTC360   \n",
       "2  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...       265     CTC360   \n",
       "3  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...        35     CTC003   \n",
       "4  era-fasp@fasp.sra.ebi.ac.uk:vol1/fastq/SRR139/...       264     CTC358   \n",
       "\n",
       "  timepoint sample_type  \n",
       "0       2.0       ctDNA  \n",
       "1       0.0       ctDNA  \n",
       "2       2.0       ctDNA  \n",
       "3       NaN        PBMC  \n",
       "4       0.0       ctDNA  \n",
       "\n",
       "[5 rows x 60 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def extract_patient_info_from_alias(run_alias):\n",
    "    \"\"\"\n",
    "    Extract patient_id, timepoint, and sample_type from run_alias.\n",
    "    Input: run_alias (str)\n",
    "    Output: dict with keys patient_id, timepoint, sample_type\n",
    "    \"\"\"\n",
    "    if pd.isna(run_alias):\n",
    "        return {\"patient_id\": None, \"timepoint\": None, \"sample_type\": None}\n",
    "    if run_alias.startswith(\"PBMC_\"):\n",
    "        patient_id = run_alias.replace(\"PBMC_\", \"\").replace(\".final.bam\", \"\")\n",
    "        return {\"patient_id\": patient_id, \"timepoint\": None, \"sample_type\": \"PBMC\"}\n",
    "    if run_alias.startswith(\"FFPE_\"):\n",
    "        patient_id = run_alias.replace(\"FFPE_\", \"\").replace(\".final.bam\", \"\")\n",
    "        return {\"patient_id\": patient_id, \"timepoint\": None, \"sample_type\": \"Tissue\"}\n",
    "    ctdna_pattern = re.match(r'^(CTC\\d+|C_fw\\d+)(?:-(\\d+))?\\.final\\.bam$', run_alias)\n",
    "    if ctdna_pattern:\n",
    "        patient_id = ctdna_pattern.group(1)\n",
    "        timepoint = 0 if ctdna_pattern.group(2) is None else int(ctdna_pattern.group(2))\n",
    "        return {\"patient_id\": patient_id, \"timepoint\": timepoint, \"sample_type\": \"ctDNA\"}\n",
    "    return {\"patient_id\": None, \"timepoint\": None, \"sample_type\": None}\n",
    "\n",
    "def add_patient_info_to_metadata(input_metadata):\n",
    "    \"\"\"\n",
    "    Add patient_id, timepoint, and sample_type columns to metadata.\n",
    "    Input: input_metadata (pd.DataFrame)\n",
    "    Output: pd.DataFrame with added patient information columns\n",
    "    \"\"\"\n",
    "    output_metadata = input_metadata.copy()\n",
    "    parsed_info = output_metadata['run_alias'].apply(extract_patient_info_from_alias)\n",
    "    parsed_df = pd.DataFrame(parsed_info.tolist())\n",
    "    columns_to_drop = [\"patient_id\", \"timepoint\", \"sample_type\"]\n",
    "    existing_columns = [col for col in columns_to_drop if col in output_metadata.columns]\n",
    "    if existing_columns:\n",
    "        output_metadata = output_metadata.drop(columns=existing_columns)\n",
    "    output_metadata = pd.concat([output_metadata, parsed_df], axis=1)\n",
    "    return output_metadata\n",
    "# Parse metadata and add patient information\n",
    "metadata = add_patient_info_to_metadata(metadata)\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c24e821e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_ctdna_samples(metadata_dataframe):\n",
    "    \"\"\"\n",
    "    Filter metadata to only ctDNA samples.\n",
    "    Input: metadata_dataframe (pd.DataFrame)\n",
    "    Output: pd.DataFrame with ctDNA samples and integer timepoints\n",
    "    \"\"\"\n",
    "    ctdna_samples = metadata_dataframe[metadata_dataframe[\"sample_type\"] == \"ctDNA\"].copy()\n",
    "    ctdna_samples[\"timepoint\"] = ctdna_samples[\"timepoint\"].astype(int)\n",
    "    return ctdna_samples\n",
    "# Filter to ctDNA samples only\n",
    "ctdna = filter_ctdna_samples(metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "131fb4d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "patient_id",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "timepoints",
         "rawType": "object",
         "type": "unknown"
        },
        {
         "name": "n_timepoints",
         "rawType": "int64",
         "type": "integer"
        }
       ],
       "ref": "0f833e48-643b-4199-994a-914f67564c2c",
       "rows": [
        [
         "0",
         "CTC002",
         "[0, 2, 5]",
         "3"
        ],
        [
         "29",
         "CTC258",
         "[0, 2, 5]",
         "3"
        ],
        [
         "30",
         "CTC259",
         "[0, 3, 6]",
         "3"
        ],
        [
         "31",
         "CTC267",
         "[0, 2, 3]",
         "3"
        ],
        [
         "32",
         "CTC271",
         "[0, 2, 3]",
         "3"
        ],
        [
         "33",
         "CTC276",
         "[0, 2, 11]",
         "3"
        ],
        [
         "34",
         "CTC277",
         "[0, 2, 6]",
         "3"
        ],
        [
         "35",
         "CTC281",
         "[0, 2, 3]",
         "3"
        ],
        [
         "36",
         "CTC292",
         "[0, 2, 10]",
         "3"
        ],
        [
         "37",
         "CTC309",
         "[0, 2, 9]",
         "3"
        ]
       ],
       "shape": {
        "columns": 3,
        "rows": 10
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>patient_id</th>\n",
       "      <th>timepoints</th>\n",
       "      <th>n_timepoints</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTC002</td>\n",
       "      <td>[0, 2, 5]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CTC258</td>\n",
       "      <td>[0, 2, 5]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>CTC259</td>\n",
       "      <td>[0, 3, 6]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>CTC267</td>\n",
       "      <td>[0, 2, 3]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>CTC271</td>\n",
       "      <td>[0, 2, 3]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>CTC276</td>\n",
       "      <td>[0, 2, 11]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>CTC277</td>\n",
       "      <td>[0, 2, 6]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>CTC281</td>\n",
       "      <td>[0, 2, 3]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>CTC292</td>\n",
       "      <td>[0, 2, 10]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>CTC309</td>\n",
       "      <td>[0, 2, 9]</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   patient_id  timepoints  n_timepoints\n",
       "0      CTC002   [0, 2, 5]             3\n",
       "29     CTC258   [0, 2, 5]             3\n",
       "30     CTC259   [0, 3, 6]             3\n",
       "31     CTC267   [0, 2, 3]             3\n",
       "32     CTC271   [0, 2, 3]             3\n",
       "33     CTC276  [0, 2, 11]             3\n",
       "34     CTC277   [0, 2, 6]             3\n",
       "35     CTC281   [0, 2, 3]             3\n",
       "36     CTC292  [0, 2, 10]             3\n",
       "37     CTC309   [0, 2, 9]             3"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def find_eligible_patients(ctdna_dataframe):\n",
    "    \"\"\"\n",
    "    Find patients with at least 3 timepoints including baseline (0).\n",
    "    Input: ctdna_dataframe (pd.DataFrame)\n",
    "    Output: pd.DataFrame with eligible patients and their timepoints\n",
    "    \"\"\"\n",
    "    eligible_patients_list = []\n",
    "    for patient_id, patient_group in ctdna_dataframe.groupby(\"patient_id\"):\n",
    "        timepoints = sorted(patient_group[\"timepoint\"].dropna().unique())\n",
    "        timepoints = [int(tp) for tp in timepoints]\n",
    "        has_baseline = 0 in timepoints\n",
    "        has_followup = any(tp > 0 for tp in timepoints)\n",
    "        has_three_timepoints = len(timepoints) >= 3\n",
    "        if has_baseline and has_followup and has_three_timepoints:\n",
    "            eligible_patients_list.append((patient_id, timepoints))\n",
    "    \n",
    "    eligible_df = pd.DataFrame(eligible_patients_list, columns=[\"patient_id\", \"timepoints\"])\n",
    "    eligible_df[\"n_timepoints\"] = eligible_df[\"timepoints\"].apply(len)\n",
    "    return eligible_df\n",
    "# Find eligible patients\n",
    "eligible_df = find_eligible_patients(ctdna)\n",
    "eligible_df.sort_values(\"n_timepoints\", ascending=True).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "416465d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "run_accession",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "run_alias",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "timepoint",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "tp_label",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "55532285-fdca-412b-bc2f-5311b1a10bc2",
       "rows": [
        [
         "108",
         "SRR13973811",
         "CTC030.final.bam",
         "0",
         "pre_treatment"
        ],
        [
         "110",
         "SRR13973813",
         "CTC030-2.final.bam",
         "2",
         "during_treatment"
        ],
        [
         "109",
         "SRR13973812",
         "CTC030-6.final.bam",
         "6",
         "post_treatment"
        ]
       ],
       "shape": {
        "columns": 4,
        "rows": 3
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>run_accession</th>\n",
       "      <th>run_alias</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>tp_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>SRR13973811</td>\n",
       "      <td>CTC030.final.bam</td>\n",
       "      <td>0</td>\n",
       "      <td>pre_treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>SRR13973813</td>\n",
       "      <td>CTC030-2.final.bam</td>\n",
       "      <td>2</td>\n",
       "      <td>during_treatment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>SRR13973812</td>\n",
       "      <td>CTC030-6.final.bam</td>\n",
       "      <td>6</td>\n",
       "      <td>post_treatment</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    run_accession           run_alias  timepoint          tp_label\n",
       "108   SRR13973811    CTC030.final.bam          0     pre_treatment\n",
       "110   SRR13973813  CTC030-2.final.bam          2  during_treatment\n",
       "109   SRR13973812  CTC030-6.final.bam          6    post_treatment"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def annotate_clinical_timepoints(patient_dataframe):\n",
    "    \"\"\"\n",
    "    Annotate timepoints with clinical phase labels.\n",
    "    Input: patient_dataframe (pd.DataFrame)\n",
    "    Output: pd.DataFrame with added tp_label column\n",
    "    \"\"\"\n",
    "    output_df = patient_dataframe.copy()\n",
    "    output_df[\"timepoint\"] = output_df[\"timepoint\"].astype(int)\n",
    "    unique_timepoints = sorted(output_df[\"timepoint\"].unique())\n",
    "    baseline_timepoint = 0\n",
    "    first_followup_timepoint = min(tp for tp in unique_timepoints if tp > 0)\n",
    "    last_timepoint = max(tp for tp in unique_timepoints if tp > 0)\n",
    "    def get_clinical_label(timepoint_value):\n",
    "        if timepoint_value == baseline_timepoint:\n",
    "            return \"pre_treatment\"\n",
    "        elif timepoint_value == first_followup_timepoint:\n",
    "            return \"during_treatment\"\n",
    "        elif timepoint_value == last_timepoint:\n",
    "            return \"post_treatment\"\n",
    "        return None\n",
    "    output_df[\"tp_label\"] = output_df[\"timepoint\"].apply(get_clinical_label)\n",
    "    return output_df.sort_values(\"timepoint\")\n",
    "# Select and annotate one patient\n",
    "selected_patient_id = \"CTC030\"\n",
    "patient_data = ctdna[ctdna[\"patient_id\"] == selected_patient_id]\n",
    "patient_data_annotated = annotate_clinical_timepoints(patient_data)\n",
    "# Display results\n",
    "patient_data_annotated[[\"run_accession\", \"run_alias\", \"timepoint\", \"tp_label\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c87164dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping CTC030_pre_treatment_R1.fastq.gz (already exists)\n",
      "Skipping CTC030_pre_treatment_R2.fastq.gz (already exists)\n",
      "Skipping CTC030_during_treatment_R1.fastq.gz (already exists)\n",
      "Skipping CTC030_during_treatment_R2.fastq.gz (already exists)\n",
      "Skipping CTC030_post_treatment_R1.fastq.gz (already exists)\n",
      "Skipping CTC030_post_treatment_R2.fastq.gz (already exists)\n"
     ]
    }
   ],
   "source": [
    "def download_single_fastq_file(download_url, output_file_path):\n",
    "    \"\"\"\n",
    "    Download a FASTQ file if it doesn't already exist.\n",
    "    Input: download_url (str), output_file_path (Path)\n",
    "    Output: bool (True if downloaded, False if skipped)\n",
    "    \"\"\"\n",
    "    if output_file_path.exists():\n",
    "        print(f\"Skipping {output_file_path.name} (already exists)\")\n",
    "        return False\n",
    "    print(f\"Downloading {output_file_path.name}...\")\n",
    "    run_cmd([\"wget\", \"-O\", str(output_file_path), download_url])\n",
    "    return True\n",
    "\n",
    "def download_fastq_files_for_patient(patient_annotated_dataframe, patient_identifier, \n",
    "                                      data_directory, timepoints_to_include=None):\n",
    "    \"\"\"\n",
    "    Download paired FASTQ files (R1 and R2) for a patient.\n",
    "    Input: patient_annotated_dataframe (pd.DataFrame), patient_identifier (str), \n",
    "           data_directory (Path), timepoints_to_include (list, optional)\n",
    "    Output: Path (directory containing downloaded files)\n",
    "    \"\"\"\n",
    "    if timepoints_to_include is None:\n",
    "        timepoints_to_include = [\"pre_treatment\", \"during_treatment\", \"post_treatment\"]\n",
    "    patient_directory = data_directory / patient_identifier\n",
    "    patient_directory.mkdir(parents=True, exist_ok=True)\n",
    "    for _, sample_row in patient_annotated_dataframe.iterrows():\n",
    "        timepoint_label = sample_row[\"tp_label\"]\n",
    "        if timepoint_label not in timepoints_to_include:\n",
    "            continue\n",
    "        r1_url = sample_row[\"ena_fastq_http_1\"]\n",
    "        r2_url = sample_row[\"ena_fastq_http_2\"]\n",
    "        if pd.notna(r1_url) and r1_url != \"\":\n",
    "            r1_file_path = patient_directory / f\"{patient_identifier}_{timepoint_label}_R1.fastq.gz\"\n",
    "            download_single_fastq_file(r1_url, r1_file_path)\n",
    "        if pd.notna(r2_url) and r2_url != \"\":\n",
    "            r2_file_path = patient_directory / f\"{patient_identifier}_{timepoint_label}_R2.fastq.gz\"\n",
    "            download_single_fastq_file(r2_url, r2_file_path)\n",
    "    return patient_directory\n",
    "\n",
    "# Download FASTQ files for the selected patient\n",
    "patient_data_directory = download_fastq_files_for_patient(\n",
    "    patient_annotated_dataframe=patient_data_annotated,\n",
    "    patient_identifier=selected_patient_id,\n",
    "    data_directory=data_dir\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4e5e8d00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'pre_treatment': {'R2': PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/CTC030/CTC030_pre_treatment_R2.fastq.gz'),\n",
       "  'R1': PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/CTC030/CTC030_pre_treatment_R1.fastq.gz')},\n",
       " 'post_treatment': {'R1': PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/CTC030/CTC030_post_treatment_R1.fastq.gz'),\n",
       "  'R2': PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/CTC030/CTC030_post_treatment_R2.fastq.gz')},\n",
       " 'during_treatment': {'R1': PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/CTC030/CTC030_during_treatment_R1.fastq.gz'),\n",
       "  'R2': PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/CTC030/CTC030_during_treatment_R2.fastq.gz')}}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def organize_fastq_files_by_timepoint(patient_directory, patient_identifier):\n",
    "    \"\"\"\n",
    "    Organize FASTQ files into nested dictionary by timepoint.\n",
    "    Input: patient_directory (Path), patient_identifier (str)\n",
    "    Output: dict with structure {timepoint: {'R1': Path, 'R2': Path}}\n",
    "    \"\"\"\n",
    "    organized_fastq_paths = {}\n",
    "    filename_pattern = re.compile(\n",
    "        rf\"^{re.escape(patient_identifier)}_(.+)_R([12])\\.fastq\\.gz$\"\n",
    "    )\n",
    "    for file_path in patient_directory.iterdir():\n",
    "        if not file_path.name.endswith(\".fastq.gz\"):\n",
    "            continue\n",
    "        match = filename_pattern.match(file_path.name)\n",
    "        if match is None:\n",
    "            continue\n",
    "        timepoint_label = match.group(1)\n",
    "        read_number = match.group(2)\n",
    "        if timepoint_label not in organized_fastq_paths:\n",
    "            organized_fastq_paths[timepoint_label] = {}\n",
    "        organized_fastq_paths[timepoint_label][f\"R{read_number}\"] = file_path\n",
    "    return organized_fastq_paths\n",
    "\n",
    "# Organize FASTQ files for alignment\n",
    "fastq_paths = organize_fastq_files_by_timepoint(\n",
    "    patient_directory=patient_data_directory,\n",
    "    patient_identifier=selected_patient_id\n",
    ")\n",
    "fastq_paths"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a51e8a51",
   "metadata": {},
   "source": [
    "### Perform alignment\n",
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8232b4f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BWA index already exists.\n",
      "Skipping CTC030_pre_treatment.sorted.bam (already aligned)\n",
      "Skipping CTC030_post_treatment.sorted.bam (already aligned)\n",
      "Skipping CTC030_during_treatment.sorted.bam (already aligned)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'pre_treatment': PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/bam/CTC030/CTC030_pre_treatment.sorted.bam'),\n",
       " 'post_treatment': PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/bam/CTC030/CTC030_post_treatment.sorted.bam'),\n",
       " 'during_treatment': PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/bam/CTC030/CTC030_during_treatment.sorted.bam')}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def build_bwa_index(reference_fasta_path):\n",
    "    \"\"\"\n",
    "    Build BWA index for reference genome.\n",
    "    Input: reference_fasta_path (Path or str)\n",
    "    Output: None\n",
    "    \"\"\"\n",
    "    reference_fasta_path = Path(reference_fasta_path)\n",
    "    index_extensions = [\".amb\", \".ann\", \".bwt\", \".pac\", \".sa\"]\n",
    "    index_file_paths = [\n",
    "        reference_fasta_path.with_suffix(reference_fasta_path.suffix + ext)\n",
    "        for ext in index_extensions\n",
    "    ]\n",
    "    if all(index_file.exists() for index_file in index_file_paths):\n",
    "        print(\"BWA index already exists.\")\n",
    "        return\n",
    "    print(\"Building BWA index...\")\n",
    "    run_cmd([\"bwa\", \"index\", str(reference_fasta_path)])\n",
    "\n",
    "def align_reads_to_reference(sample_label, forward_reads_path, reverse_reads_path,\n",
    "                              reference_fasta_path, output_directory, patient_identifier):\n",
    "    \"\"\"\n",
    "    Align paired-end reads to reference genome using BWA-MEM.\n",
    "    Input: sample_label (str), forward_reads_path (Path), reverse_reads_path (Path),\n",
    "           reference_fasta_path (Path), output_directory (Path), patient_identifier (str)\n",
    "    Output: Path to sorted BAM file\n",
    "    \"\"\"\n",
    "    output_directory = Path(output_directory)\n",
    "    output_directory.mkdir(parents=True, exist_ok=True)\n",
    "    sorted_bam_path = output_directory / f\"{patient_identifier}_{sample_label}.sorted.bam\"\n",
    "    bam_index_path = sorted_bam_path.with_suffix(\".bam.bai\")\n",
    "    if sorted_bam_path.exists() and bam_index_path.exists():\n",
    "        print(f\"Skipping {sorted_bam_path.name} (already aligned)\")\n",
    "        return sorted_bam_path\n",
    "    print(f\"Aligning {sample_label}...\")\n",
    "    bwa_mem_command = [\n",
    "        \"bwa\", \"mem\", \"-t\", \"10\",\n",
    "        str(reference_fasta_path),\n",
    "        str(forward_reads_path),\n",
    "        str(reverse_reads_path)\n",
    "    ]\n",
    "    process_bwa = subprocess.Popen(bwa_mem_command, stdout=subprocess.PIPE)\n",
    "    process_view = subprocess.Popen(\n",
    "        [\"samtools\", \"view\", \"-b\", \"-@\", \"4\"],\n",
    "        stdin=process_bwa.stdout,\n",
    "        stdout=subprocess.PIPE\n",
    "    )\n",
    "    process_sort = subprocess.Popen(\n",
    "        [\"samtools\", \"sort\", \"-@\", \"4\", \"-o\", str(sorted_bam_path)],\n",
    "        stdin=process_view.stdout\n",
    "    )\n",
    "    process_sort.communicate()\n",
    "    process_bwa.stdout.close()\n",
    "    process_view.stdout.close()\n",
    "    run_cmd([\"samtools\", \"index\", \"-@\", \"4\", str(sorted_bam_path)])\n",
    "    return sorted_bam_path\n",
    "\n",
    "# Build BWA index for reference genome\n",
    "build_bwa_index(fa_ref_genome_path)\n",
    "# Align all samples\n",
    "bam_output_directory = data_dir / \"bam\" / selected_patient_id\n",
    "aligned_bam_paths = {}\n",
    "for timepoint_label, fastq_file_paths in fastq_paths.items():\n",
    "    aligned_bam_paths[timepoint_label] = align_reads_to_reference(\n",
    "        sample_label=timepoint_label,\n",
    "        forward_reads_path=fastq_file_paths[\"R1\"],\n",
    "        reverse_reads_path=fastq_file_paths[\"R2\"],\n",
    "        reference_fasta_path=fa_ref_genome_path,\n",
    "        output_directory=bam_output_directory,\n",
    "        patient_identifier=selected_patient_id\n",
    "    )\n",
    "aligned_bam_paths"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27be637d",
   "metadata": {},
   "source": [
    "### Duplicate removal\n",
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d0f60086",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping CTC030_pre_treatment.sorted.markdup.bam (already marked)\n",
      "Skipping CTC030_post_treatment.sorted.markdup.bam (already marked)\n",
      "Skipping CTC030_during_treatment.sorted.markdup.bam (already marked)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'pre_treatment': PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/bam/CTC030/CTC030_pre_treatment.sorted.markdup.bam'),\n",
       " 'post_treatment': PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/bam/CTC030/CTC030_post_treatment.sorted.markdup.bam'),\n",
       " 'during_treatment': PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/bam/CTC030/CTC030_during_treatment.sorted.markdup.bam')}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def mark_duplicate_reads(input_bam_path, output_bam_path):\n",
    "    \"\"\"\n",
    "    Mark duplicate reads in BAM file using samtools.\n",
    "    Input: input_bam_path (Path), output_bam_path (Path)\n",
    "    Output: Path to deduplicated BAM file\n",
    "    \"\"\"\n",
    "    output_bam_path = Path(output_bam_path)\n",
    "    output_bam_index = output_bam_path.with_suffix(\".bam.bai\")\n",
    "    if output_bam_path.exists() and output_bam_index.exists():\n",
    "        print(f\"Skipping {output_bam_path.name} (already marked)\")\n",
    "        return output_bam_path\n",
    "    print(f\"Marking duplicates for {input_bam_path.name}...\")\n",
    "    name_sorted_bam = str(input_bam_path).replace(\".bam\", \".name_sorted.bam\")\n",
    "    fixmate_bam = str(input_bam_path).replace(\".bam\", \".fixmate.bam\")\n",
    "    run_cmd([\"samtools\", \"sort\", \"-n\", \"-@\", \"4\", \"-o\", name_sorted_bam, str(input_bam_path)])\n",
    "    run_cmd([\"samtools\", \"fixmate\", \"-m\", \"-@\", \"4\", name_sorted_bam, fixmate_bam])\n",
    "    run_cmd([\"samtools\", \"sort\", \"-@\", \"4\", \"-o\", str(output_bam_path), fixmate_bam])\n",
    "    run_cmd([\"samtools\", \"markdup\", \"-@\", \"4\", str(output_bam_path), str(output_bam_path) + \".tmp\"])\n",
    "    run_cmd([\"mv\", str(output_bam_path) + \".tmp\", str(output_bam_path)])\n",
    "    run_cmd([\"samtools\", \"index\", \"-@\", \"4\", str(output_bam_path)])\n",
    "    Path(name_sorted_bam).unlink(missing_ok=True)\n",
    "    Path(fixmate_bam).unlink(missing_ok=True)\n",
    "    return output_bam_path\n",
    "\n",
    "deduplicated_bam_paths = {}\n",
    "for timepoint_label, bam_path in aligned_bam_paths.items():\n",
    "    output_bam = bam_path.with_name(bam_path.stem + \".markdup.bam\")\n",
    "    deduplicated_bam_paths[timepoint_label] = mark_duplicate_reads(bam_path, output_bam)\n",
    "deduplicated_bam_paths"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c74a4a0",
   "metadata": {},
   "source": [
    "### Gene Selection\n",
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "cc07c055",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KRAS: 12:25205246-25250936\n"
     ]
    }
   ],
   "source": [
    "def get_gene_coordinates_from_ensembl(gene_name):\n",
    "    \"\"\"\n",
    "    Retrieve genomic coordinates for a gene from Ensembl REST API.\n",
    "    Input: gene_name (str)\n",
    "    Output: str (genomic region in format chromosome:start-end)\n",
    "    \"\"\"\n",
    "    ensembl_api_url = f\"https://rest.ensembl.org/lookup/symbol/homo_sapiens/{gene_name}\"\n",
    "    api_response = requests.get(ensembl_api_url, headers={\"Content-Type\": \"application/json\"})\n",
    "    if not api_response.ok:\n",
    "        raise ValueError(f\"Gene '{gene_name}' not found\")\n",
    "    gene_data = api_response.json()\n",
    "    genomic_region = f\"{gene_data['seq_region_name']}:{gene_data['start']}-{gene_data['end']}\"\n",
    "    print(f\"{gene_name}: {genomic_region}\")\n",
    "    return genomic_region\n",
    "\n",
    "kras_genomic_region = get_gene_coordinates_from_ensembl(\"KRAS\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88f198c5",
   "metadata": {},
   "source": [
    "### Variant Calling\n",
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e1d01b35",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Note: none of --samples-file, --ploidy or --ploidy-file given, assuming all sites are diploid\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 250\n",
      "Note: none of --samples-file, --ploidy or --ploidy-file given, assuming all sites are diploid\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 250\n",
      "Note: none of --samples-file, --ploidy or --ploidy-file given, assuming all sites are diploid\n",
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] maximum number of reads per input file set to -d 250\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'pre_treatment': PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/vcf/CTC030/CTC030_pre_treatment_KRAS.vcf.gz'),\n",
       " 'post_treatment': PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/vcf/CTC030/CTC030_post_treatment_KRAS.vcf.gz'),\n",
       " 'during_treatment': PosixPath('/Users/nzlab-la/Desktop/circulatory_tumor_DNA_analysis/data/vcf/CTC030/CTC030_during_treatment_KRAS.vcf.gz')}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def call_variants_with_bcftools(bam_file_path, reference_fasta_path, genomic_region, output_vcf_path):\n",
    "    \"\"\"\n",
    "    Call genetic variants using bcftools mpileup and call pipeline.\n",
    "    Input: bam_file_path (Path), reference_fasta_path (Path), genomic_region (str), output_vcf_path (Path)\n",
    "    Output: None\n",
    "    \"\"\"\n",
    "    # Align reads at each position in the genomic region\n",
    "    mpileup_command = [\n",
    "        \"bcftools\", \"mpileup\",\n",
    "        \"-f\", str(reference_fasta_path),\n",
    "        \"-r\", genomic_region,\n",
    "        \"-a\", \"FORMAT/DP,FORMAT/AD\",\n",
    "        str(bam_file_path)\n",
    "    ]\n",
    "    # Identifies variants from pileup data\n",
    "    call_command = [\n",
    "        \"bcftools\", \"call\",\n",
    "        \"-mv\",\n",
    "        \"-Oz\",\n",
    "        \"-o\", str(output_vcf_path)\n",
    "    ]\n",
    "    process_mpileup = subprocess.Popen(mpileup_command, stdout=subprocess.PIPE)\n",
    "    process_call = subprocess.Popen(call_command, stdin=process_mpileup.stdout)\n",
    "    process_call.communicate()\n",
    "    process_mpileup.stdout.close()\n",
    "    run_cmd([\"bcftools\", \"index\", str(output_vcf_path)])\n",
    "\n",
    "vcf_output_directory = data_dir / \"vcf\" / selected_patient_id\n",
    "vcf_output_directory.mkdir(parents=True, exist_ok=True)\n",
    "kras_vcf_file_paths = {}\n",
    "for timepoint_label, bam_file_path in deduplicated_bam_paths.items():\n",
    "    output_vcf_file = vcf_output_directory / f\"{selected_patient_id}_{timepoint_label}_KRAS.vcf.gz\"\n",
    "    call_variants_with_bcftools(\n",
    "        bam_file_path=bam_file_path,\n",
    "        reference_fasta_path=fa_ref_genome_path,\n",
    "        genomic_region=kras_genomic_region,\n",
    "        output_vcf_path=output_vcf_file\n",
    "    )\n",
    "    kras_vcf_file_paths[timepoint_label] = output_vcf_file\n",
    "kras_vcf_file_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e2742322",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "sample_label",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "chrom",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "position",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "reference",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "variant",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "read_depth",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "allele_frequency",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "39bb670b-3add-4889-bdb5-6025e72b55c1",
       "rows": [
        [
         "0",
         "pre_treatment",
         "12",
         "25205716",
         "A",
         "T",
         "1",
         "1.0"
        ],
        [
         "1",
         "pre_treatment",
         "12",
         "25205728",
         "CTTTTTTTT",
         "CTTTTTT",
         "2",
         "1.0"
        ],
        [
         "2",
         "pre_treatment",
         "12",
         "25207204",
         "T",
         "C",
         "1",
         "1.0"
        ],
        [
         "3",
         "pre_treatment",
         "12",
         "25208712",
         "T",
         "C",
         "2",
         "1.0"
        ],
        [
         "4",
         "pre_treatment",
         "12",
         "25209531",
         "G",
         "A",
         "21",
         "1.0"
        ],
        [
         "5",
         "pre_treatment",
         "12",
         "25209618",
         "A",
         "C",
         "44",
         "0.5227"
        ],
        [
         "6",
         "pre_treatment",
         "12",
         "25209843",
         "A",
         "G",
         "182",
         "0.4088"
        ],
        [
         "7",
         "pre_treatment",
         "12",
         "25214904",
         "A",
         "C",
         "1",
         "1.0"
        ],
        [
         "8",
         "pre_treatment",
         "12",
         "25215150",
         "T",
         "C",
         "11",
         "0.3636"
        ],
        [
         "9",
         "pre_treatment",
         "12",
         "25218970",
         "C",
         "T",
         "1",
         "1.0"
        ],
        [
         "10",
         "pre_treatment",
         "12",
         "25225092",
         "G",
         "A",
         "1",
         "1.0"
        ],
        [
         "11",
         "pre_treatment",
         "12",
         "25225286",
         "TTATATATATATATATATATATATATATATATATATATATATATATATAT",
         "TTATATATATATATATATATATAT",
         "8",
         "0.625"
        ],
        [
         "12",
         "pre_treatment",
         "12",
         "25225290",
         "A",
         "T",
         "5",
         "1.0"
        ],
        [
         "13",
         "pre_treatment",
         "12",
         "25226703",
         "T",
         "C",
         "3",
         "1.0"
        ],
        [
         "14",
         "pre_treatment",
         "12",
         "25226859",
         "C",
         "T",
         "29",
         "0.5517"
        ],
        [
         "15",
         "pre_treatment",
         "12",
         "25227137",
         "T",
         "G",
         "77",
         "0.4675"
        ],
        [
         "16",
         "pre_treatment",
         "12",
         "25249453",
         "T",
         "C",
         "4",
         "0.5"
        ],
        [
         "17",
         "post_treatment",
         "12",
         "25209618",
         "A",
         "C",
         "2",
         "0.5"
        ],
        [
         "18",
         "post_treatment",
         "12",
         "25209843",
         "A",
         "G",
         "178",
         "0.3652"
        ],
        [
         "19",
         "post_treatment",
         "12",
         "25227137",
         "T",
         "G",
         "20",
         "0.4"
        ],
        [
         "20",
         "post_treatment",
         "12",
         "25228768",
         "T",
         "A",
         "1",
         "1.0"
        ],
        [
         "21",
         "post_treatment",
         "12",
         "25235963",
         "T",
         "G",
         "1",
         "1.0"
        ],
        [
         "22",
         "post_treatment",
         "12",
         "25246914",
         "CAAAAAAAAAAAAA",
         "CAAAAAAAAAAAAAA",
         "2",
         "1.0"
        ],
        [
         "23",
         "during_treatment",
         "12",
         "25209618",
         "A",
         "C",
         "1",
         "1.0"
        ],
        [
         "24",
         "during_treatment",
         "12",
         "25209843",
         "A",
         "G",
         "192",
         "0.4712"
        ],
        [
         "25",
         "during_treatment",
         "12",
         "25231904",
         "G",
         "A",
         "1",
         "1.0"
        ],
        [
         "26",
         "during_treatment",
         "12",
         "25245002",
         "T",
         "A",
         "4",
         "0.25"
        ]
       ],
       "shape": {
        "columns": 7,
        "rows": 27
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_label</th>\n",
       "      <th>chrom</th>\n",
       "      <th>position</th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>read_depth</th>\n",
       "      <th>allele_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25205716</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25205728</td>\n",
       "      <td>CTTTTTTTT</td>\n",
       "      <td>CTTTTTT</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25207204</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25208712</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25209531</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>21</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25209618</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>44</td>\n",
       "      <td>0.5227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25209843</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>182</td>\n",
       "      <td>0.4088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25214904</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25215150</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>11</td>\n",
       "      <td>0.3636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25218970</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25225092</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25225286</td>\n",
       "      <td>TTATATATATATATATATATATATATATATATATATATATATATAT...</td>\n",
       "      <td>TTATATATATATATATATATATAT</td>\n",
       "      <td>8</td>\n",
       "      <td>0.6250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25225290</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>5</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25226703</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25226859</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>29</td>\n",
       "      <td>0.5517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25227137</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>77</td>\n",
       "      <td>0.4675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25249453</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>4</td>\n",
       "      <td>0.5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>post_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25209618</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>0.5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>post_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25209843</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>178</td>\n",
       "      <td>0.3652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>post_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25227137</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>20</td>\n",
       "      <td>0.4000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>post_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25228768</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>post_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25235963</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>post_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25246914</td>\n",
       "      <td>CAAAAAAAAAAAAA</td>\n",
       "      <td>CAAAAAAAAAAAAAA</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>during_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25209618</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>during_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25209843</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>192</td>\n",
       "      <td>0.4712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>during_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25231904</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>during_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25245002</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "      <td>0.2500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample_label chrom  position  \\\n",
       "0      pre_treatment    12  25205716   \n",
       "1      pre_treatment    12  25205728   \n",
       "2      pre_treatment    12  25207204   \n",
       "3      pre_treatment    12  25208712   \n",
       "4      pre_treatment    12  25209531   \n",
       "5      pre_treatment    12  25209618   \n",
       "6      pre_treatment    12  25209843   \n",
       "7      pre_treatment    12  25214904   \n",
       "8      pre_treatment    12  25215150   \n",
       "9      pre_treatment    12  25218970   \n",
       "10     pre_treatment    12  25225092   \n",
       "11     pre_treatment    12  25225286   \n",
       "12     pre_treatment    12  25225290   \n",
       "13     pre_treatment    12  25226703   \n",
       "14     pre_treatment    12  25226859   \n",
       "15     pre_treatment    12  25227137   \n",
       "16     pre_treatment    12  25249453   \n",
       "17    post_treatment    12  25209618   \n",
       "18    post_treatment    12  25209843   \n",
       "19    post_treatment    12  25227137   \n",
       "20    post_treatment    12  25228768   \n",
       "21    post_treatment    12  25235963   \n",
       "22    post_treatment    12  25246914   \n",
       "23  during_treatment    12  25209618   \n",
       "24  during_treatment    12  25209843   \n",
       "25  during_treatment    12  25231904   \n",
       "26  during_treatment    12  25245002   \n",
       "\n",
       "                                            reference  \\\n",
       "0                                                   A   \n",
       "1                                           CTTTTTTTT   \n",
       "2                                                   T   \n",
       "3                                                   T   \n",
       "4                                                   G   \n",
       "5                                                   A   \n",
       "6                                                   A   \n",
       "7                                                   A   \n",
       "8                                                   T   \n",
       "9                                                   C   \n",
       "10                                                  G   \n",
       "11  TTATATATATATATATATATATATATATATATATATATATATATAT...   \n",
       "12                                                  A   \n",
       "13                                                  T   \n",
       "14                                                  C   \n",
       "15                                                  T   \n",
       "16                                                  T   \n",
       "17                                                  A   \n",
       "18                                                  A   \n",
       "19                                                  T   \n",
       "20                                                  T   \n",
       "21                                                  T   \n",
       "22                                     CAAAAAAAAAAAAA   \n",
       "23                                                  A   \n",
       "24                                                  A   \n",
       "25                                                  G   \n",
       "26                                                  T   \n",
       "\n",
       "                     variant  read_depth  allele_frequency  \n",
       "0                          T           1            1.0000  \n",
       "1                    CTTTTTT           2            1.0000  \n",
       "2                          C           1            1.0000  \n",
       "3                          C           2            1.0000  \n",
       "4                          A          21            1.0000  \n",
       "5                          C          44            0.5227  \n",
       "6                          G         182            0.4088  \n",
       "7                          C           1            1.0000  \n",
       "8                          C          11            0.3636  \n",
       "9                          T           1            1.0000  \n",
       "10                         A           1            1.0000  \n",
       "11  TTATATATATATATATATATATAT           8            0.6250  \n",
       "12                         T           5            1.0000  \n",
       "13                         C           3            1.0000  \n",
       "14                         T          29            0.5517  \n",
       "15                         G          77            0.4675  \n",
       "16                         C           4            0.5000  \n",
       "17                         C           2            0.5000  \n",
       "18                         G         178            0.3652  \n",
       "19                         G          20            0.4000  \n",
       "20                         A           1            1.0000  \n",
       "21                         G           1            1.0000  \n",
       "22           CAAAAAAAAAAAAAA           2            1.0000  \n",
       "23                         C           1            1.0000  \n",
       "24                         G         192            0.4712  \n",
       "25                         A           1            1.0000  \n",
       "26                         A           4            0.2500  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def parse_vcf_to_dataframe(vcf_file_path, timepoint_label):\n",
    "    \"\"\"\n",
    "    Parse VCF file and extract variant information into DataFrame.\n",
    "    Input: vcf_file_path (Path), timepoint_label (str)\n",
    "    Output: pd.DataFrame with variant data including allele frequencies\n",
    "    \"\"\"\n",
    "    variant_rows = []\n",
    "    with pysam.VariantFile(str(vcf_file_path)) as vcf_file:\n",
    "        sample_name = list(vcf_file.header.samples)[0]\n",
    "        for variant_record in vcf_file:\n",
    "            sample_format = variant_record.samples[sample_name]\n",
    "            read_depth = sample_format.get(\"DP\", None)\n",
    "            allele_depths = sample_format.get(\"AD\", None)\n",
    "            allele_frequency = None\n",
    "            if allele_depths and isinstance(allele_depths, (list, tuple)) and len(allele_depths) > 1:\n",
    "                reference_read_count = allele_depths[0]\n",
    "                alternate_read_count = sum(allele_depths[1:])\n",
    "                total_read_count = reference_read_count + alternate_read_count\n",
    "                if total_read_count > 0:\n",
    "                    allele_frequency = np.round(alternate_read_count / total_read_count, 4)\n",
    "            variant_rows.append({\n",
    "                \"sample_label\": timepoint_label,\n",
    "                \"chrom\": variant_record.chrom,\n",
    "                \"position\": variant_record.pos,\n",
    "                \"reference\": variant_record.ref,\n",
    "                \"variant\": \",\".join(variant_record.alts),\n",
    "                \"read_depth\": read_depth,\n",
    "                \"allele_frequency\": allele_frequency,\n",
    "            })\n",
    "    return pd.DataFrame(variant_rows)\n",
    "\n",
    "variant_dataframes = []\n",
    "for timepoint_label, vcf_file_path in kras_vcf_file_paths.items():\n",
    "    variant_dataframes.append(parse_vcf_to_dataframe(vcf_file_path, timepoint_label))\n",
    "\n",
    "kras_variants_dataframe = pd.concat(variant_dataframes, ignore_index=True)\n",
    "kras_variants_dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba58e929",
   "metadata": {},
   "source": [
    "### Selecting a candidate variant for diagnossis assays development\n",
    "___"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d2c1b951",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "sample_label",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "chrom",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "position",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "reference",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "variant",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "read_depth",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "allele_frequency",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "4b906634-00d3-4623-beb7-38a563d85332",
       "rows": [
        [
         "4",
         "pre_treatment",
         "12",
         "25209531",
         "G",
         "A",
         "21",
         "1.0"
        ],
        [
         "5",
         "pre_treatment",
         "12",
         "25209618",
         "A",
         "C",
         "44",
         "0.5227"
        ],
        [
         "6",
         "pre_treatment",
         "12",
         "25209843",
         "A",
         "G",
         "182",
         "0.4088"
        ],
        [
         "18",
         "post_treatment",
         "12",
         "25209843",
         "A",
         "G",
         "178",
         "0.3652"
        ],
        [
         "24",
         "during_treatment",
         "12",
         "25209843",
         "A",
         "G",
         "192",
         "0.4712"
        ],
        [
         "14",
         "pre_treatment",
         "12",
         "25226859",
         "C",
         "T",
         "29",
         "0.5517"
        ],
        [
         "15",
         "pre_treatment",
         "12",
         "25227137",
         "T",
         "G",
         "77",
         "0.4675"
        ],
        [
         "19",
         "post_treatment",
         "12",
         "25227137",
         "T",
         "G",
         "20",
         "0.4"
        ]
       ],
       "shape": {
        "columns": 7,
        "rows": 8
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_label</th>\n",
       "      <th>chrom</th>\n",
       "      <th>position</th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>read_depth</th>\n",
       "      <th>allele_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25209531</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>21</td>\n",
       "      <td>1.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25209618</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>44</td>\n",
       "      <td>0.5227</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25209843</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>182</td>\n",
       "      <td>0.4088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>post_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25209843</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>178</td>\n",
       "      <td>0.3652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>during_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25209843</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>192</td>\n",
       "      <td>0.4712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25226859</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>29</td>\n",
       "      <td>0.5517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25227137</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>77</td>\n",
       "      <td>0.4675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>post_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25227137</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>20</td>\n",
       "      <td>0.4000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample_label chrom  position reference variant  read_depth  \\\n",
       "4      pre_treatment    12  25209531         G       A          21   \n",
       "5      pre_treatment    12  25209618         A       C          44   \n",
       "6      pre_treatment    12  25209843         A       G         182   \n",
       "18    post_treatment    12  25209843         A       G         178   \n",
       "24  during_treatment    12  25209843         A       G         192   \n",
       "14     pre_treatment    12  25226859         C       T          29   \n",
       "15     pre_treatment    12  25227137         T       G          77   \n",
       "19    post_treatment    12  25227137         T       G          20   \n",
       "\n",
       "    allele_frequency  \n",
       "4             1.0000  \n",
       "5             0.5227  \n",
       "6             0.4088  \n",
       "18            0.3652  \n",
       "24            0.4712  \n",
       "14            0.5517  \n",
       "15            0.4675  \n",
       "19            0.4000  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filtering for read depth and allele frequency\n",
    "filtered_variants = kras_variants_dataframe[\n",
    "    (kras_variants_dataframe[\"read_depth\"] >= 20) &      # Minimum 20x coverage\n",
    "    (kras_variants_dataframe[\"allele_frequency\"] >= 0.1)       # Minimum 10% allele frequency\n",
    "]\n",
    "filtered_variants\n",
    "# Sorting based on chromosome and position to determine time of occurrence\n",
    "filtered_variants = filtered_variants.sort_values(by=[\"position\"])\n",
    "filtered_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6a3465cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "sample_label",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "chrom",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "position",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "reference",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "variant",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "read_depth",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "allele_frequency",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "7ac000f8-d57e-461f-a999-eb073e2fc473",
       "rows": [
        [
         "6",
         "pre_treatment",
         "12",
         "25209843",
         "A",
         "G",
         "182",
         "0.4088"
        ],
        [
         "18",
         "post_treatment",
         "12",
         "25209843",
         "A",
         "G",
         "178",
         "0.3652"
        ],
        [
         "24",
         "during_treatment",
         "12",
         "25209843",
         "A",
         "G",
         "192",
         "0.4712"
        ]
       ],
       "shape": {
        "columns": 7,
        "rows": 3
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_label</th>\n",
       "      <th>chrom</th>\n",
       "      <th>position</th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>read_depth</th>\n",
       "      <th>allele_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>pre_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25209843</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>182</td>\n",
       "      <td>0.4088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>post_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25209843</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>178</td>\n",
       "      <td>0.3652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>during_treatment</td>\n",
       "      <td>12</td>\n",
       "      <td>25209843</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>192</td>\n",
       "      <td>0.4712</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample_label chrom  position reference variant  read_depth  \\\n",
       "6      pre_treatment    12  25209843         A       G         182   \n",
       "18    post_treatment    12  25209843         A       G         178   \n",
       "24  during_treatment    12  25209843         A       G         192   \n",
       "\n",
       "    allele_frequency  \n",
       "6             0.4088  \n",
       "18            0.3652  \n",
       "24            0.4712  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# selecting variant position 25209843\n",
    "selected_variant = kras_variants_dataframe[kras_variants_dataframe[\"position\"] == 25209843].copy()\n",
    "selected_variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdbec78c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAk0AAAGGCAYAAABmPbWyAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjgsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvwVt1zgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAZppJREFUeJzt3Xl8U1XaB/DfTdqkSTe6U7a2CCgICBQXUMQFqOACDCqMCzAiisvIIi7oO6LMKI6K4Dgig4q4L4gbyig4iqyKIMUFkK1QlkJXmq5J2zzvH7XXpEkhKelN2/v7fj6F9uTm5pzkPsmTc885VxERARERERGdlCHYFSAiIiJqCZg0EREREfmASRMRERGRD5g0EREREfmASRMRERGRD5g0EREREfmASRMRERGRD5g0EREREfmASRMRERGRD5g0NVOPPvooFEWBoiiYOHGi221LliyB0WiEoigICwvD559/DgCYOHGiep+6n9DQULRr1w5jxozB5s2bG3y88vJyREVFqfczGAw4cOCA121FBK+88goGDhyIqKgomEwmJCYmomfPnhg3bhw+++yzk7btjTfeUB/nvPPO87rNvn371G0iIyNRVlZ20n3665JLLlH331A7A23p0qV49NFH8eijj+LEiRN+39/1eVMUBZdeeqnX7VyPnaVLl6rlqampanljBWIfDal/7Nb/WbBgQcAfk/yzYsUKjB49Gu3bt4fZbEZ0dDT69euHhx9+GMeOHQt29XySm5uLqVOnomvXrggLC0N4eDg6dOiACy+8EHfccQdycnLctq+LWR5/BAAQapZmz54tAASATJgwQS1/4YUXRFEUASAWi0W+/PJL9bYJEyao9/H2YzabZevWrV4f7/XXX/fY/tFHH/W67dSpU0/6OHfddddJ21ZWViaRkZHq9rt27Tpp+ydOnOjDM+afwYMHq/vPysoK+P6b4jEvu+wyt+dZURSv+3F97l599VW1PCUlRS1vrEDsoyEnO6YAyPz58wP+mOQbh8Mh119//Ulfn9jYWPnmm2+CXdWTKioqktTU1JO244cffnC7T115SkpKcCpNzQp7mlqQ+fPn46677oKIIDw8HJ9//jmGDRvmddvZs2dDRJCVlYW+ffsCAOx2OxYtWuR1e9ceibpehNdeew1S79KEeXl5eP755wEAnTp1wtatW1FZWYljx45h/fr1eOCBB5CSknLSdlitVlx33XXq32+88YbHNm+++ab6e/2ettNRXl4OAFizZg1EBCKC1NTUgO2/qWRnZ2PNmjUA/nh9RASvvfZaEGvVdLKystTXp+5n2rRpJ71P3WtLgffggw/i/fffBwAkJiZixYoVqKioQHZ2NiZMmAAAKCwsxOjRo3H48OGg1dPpdMJutzd4+yuvvKL2LE+cOBFHjx5FZWUl9u7diw8++ABjx45FWFiYRrWlFilY2RqdXP2epieeeEL9OzIyUtatW+dxH9eeptmzZ6vlzz33nFo+bNgwj/sdPHhQDAaDAJC0tDTJyMhQt1+zZo3bths3blRvGzJkSKPbt3btWrdvcE6nU71tw4YN6m1paWnidDqloqJCJk6cKOecc47Ex8dLaGioWK1W6dWrl/ztb3+T0tJSt/277nvjxo1y8cUXi9VqlcGDB4tIw70+U6dOlXPPPVcSExPFZDKJxWKRbt26ydSpUyUvL8/tMVx7XX777TcZNWqUREVFSXx8vFx//fVy/PhxERH55ptvTvrN1pdepzlz5qjb33777W6vl+tzJ+J/T5PD4ZD58+fLueeeKxEREWIymaRbt27ywAMPSHFxcYNtbuw+GuLrc+Jahx07dsiVV14pkZGRkpqaqm6zf/9+ue222yQtLU1MJpNERkbKoEGD5P333/fY38GDB2XMmDESGRkpUVFRcv3110t2drbXHoaGnlvX19i1Z9ifutTfx5tvvim9e/eWsLAw6dKlizz77LMer3VhYaE89NBD0qtXL7FarRIWFiZnnHGG3H777SIi8pe//EXd58qVK93u+9e//lW97eOPP27w+c7NzRWTyaRu+8EHH7jd7nQ6pUePHurt06dPFxGR0aNHq2UbNmxwu8+tt96q3rZixQq1/Ntvv5VRo0ZJUlKShIaGSkJCgvzpT3+SLVu2uN3f9XVYtGiRPPDAA9KhQwcxGAwn7e267bbb1Pu9+eabDW5X/zHq/9QdE0ePHpUbbrhBevToIbGxsRISEiKRkZGSnp4u8+bNk6qqKrd9FhUVya233iqxsbFitVolIyNDfvnll4DE1YYNGyQjI0Pi4uLEaDRKmzZtpEePHnLTTTfJvn37TtpW8h2TpmbKNWBjY2PV36Ojo2XTpk1e79NQ0rRgwQK1/KabbvK4n+sH8oMPPiivvPJKg6fGDh8+7Pbm0bt3b5kxY4a89957cuTIEb/a2KVLF3U/3377rVo+ZcoUtbzuFGFRUdFJE4/6yWBdudVqFYvFov59qqQpOjq6wcc4++yzxeFwqNu6vtHFxMQ0WKdAJE2uz9WuXbtk0KBBDSa2/iRNlZWVbs9F/Z/u3btLYWFhQPfRkMYkTfHx8R4fZJs3b3Y7/Vv/58EHH1T3VVhYKJ06dfLYpmPHjgFJmvypi+s+vB1PAOTtt99Wt8/KynKrp+tPdHS0iIj8+uuv6un84cOHq/etqqqSxMREASAdOnSQ6urqBp/vd955x61eNTU1HtvMmzfP7fUWEfnvf/+rlk2ZMkXdtrKyUo0z18deuHChWtf6P6GhoW7Jlevr4HoMADhp0vSPf/xD3c5kMslVV10lc+fOla+//lrKy8vdtvUladq2bdtJY/u2225T9+dwOOS8887z2KZNmzZux4jr8+RrXB06dEgiIiIa3Hb16tUNPifkHyZNzVRDAbt+/foG7+MtacrKypK+ffuq5V988YXH/Vw/kLdt2yaFhYUSGhoqACQiIsKjF+eGG27wWjdFUeTSSy+VX375xac2uiZrt956q4iI2O12NUlUFEX2798vIrVvIG+99Zbs27dPSkpKxOFwyN69e6VPnz7qPn766Sd13671uvTSS2X37t1SVlYmv/76q4g0nDS9/fbbsmvXLjlx4oRUVVXJ4cOH5YorrlC3/fTTT9VtXT+8L7nkEjl06JDs2rVL/TACIDk5Oer2jR3T5Nord84554iIyPPPP6+W1U9s/UmannnmGbVs1qxZUlBQIGVlZfLPf/5TLZ85c2ZA99GQk334uD5nrnXo1auXbNu2TcrLy9XXv2fPnuqH0VdffSWVlZWSnZ2tJpqKosjPP/8sIiJ/+9vf1H2de+65sn//fjl+/LiMHDnS4wPyZM9tQ0mTP3Wpn1w/++yzUlxc7PZaZ2RkqPu++uqr1fILLrhAfvzxRykrK5MdO3bInDlz1O2uvPJK9bH27t0rIiIrV65U79vQ2MU6rq9j3759vW7z8ccfq9tYrVYREampqVHHD8XExIjdbhcRkffff9/jferw4cNiNpsFgPTr10927twpdrtdtmzZIgkJCQJAkpOT1Z4b19chNDRUli5dKjabTQ4cOCD5+fkNtiUnJ0fi4uK8Hl8REREyY8YMqaysdLuPt+OgzvHjx2X58uVy8OBBKSsrE7vdLj/99JN06NBBAIjRaJSioiIREVmyZInbvureaydNmuRWjzr+xNXy5cvVsqeffloqKyulsLBQtm7dKnPnzlWPMTp9TJqaqYaSpqFDh3p8I6pzsoHgCQkJ8vLLL3vcx/UDuVu3bmr5iBEj1PKlS5e63ae6ulqeffZZOfvss70+VkpKikei5c3BgwfVb5bR0dFSUVEhH374oVsi4uqVV16Riy66SGJiYtTTU64/7777rrqta3l2drbHYzeUwHz00UcyZMgQiY+PF6PR6PEYTz75pLqt64f39u3b1fIxY8ao5a69go1Nmm655Rb1fk888YSI1L751z0H9RNbf5KmCy+88JTJSs+ePQO6j4acah/ekqa1a9e67WPPnj2n3A8AeeaZZ0RE5Pzzz1fLXHsyfvvtN68flv4kTf7WxXUfrslJSUmJWn7mmWeKiEhFRYWEhISo5QcOHGjweV2zZo26Xd2ps7ovPkajUQ4fPnzS1+XJJ5/0Wi9XH330kbpNXdIkIvL3v/9dLV++fLmIiFx11VXqY9fF5ksvveTTc1V3ms71dbjllltOWv/6Dh06JBMnTnTrwXf9uffee92293Yc1KmqqpJnnnlGzj33XImKivLaU/bdd9+JiMi4cePUsueee07dR0lJidtrWcefuHLt8erbt6888sgj8tZbb8lPP/3kcUqXTg8HgrcA6enpaNOmDQBg9erVuPLKK/2egm+3270OlHUdAD5gwABkZmYiMzMT/fr187oNABiNRkyfPh2//PILDh06hHfffRfjxo1Tbz948CA2bdp0yjp16tRJnTZfXFyMFStWuA0Kdx0APm/ePEyaNAnr169HUVERnE6nx/4qKio8yhISEtCxY8dT1gUAPvjgA4wePRpfffUV8vPzUVNT49NjAED37t3V38PDw9XfKysrfXrshpSXl2PZsmXq32eeeSYyMzNx7Ngx9OzZEwBQWlqKDz74oFH7P378+Cm3yc/Pb/J91OdtILi3Afvp6el+18W1PgUFBWqZ63FyqskMANwmSVRXV3vc7m9dXJ3qeCooKFAfMzIy8qT1HTx4MM4991wAwKuvvorc3Fx88sknAICrr74a7du3P2n9XPd94MABr7G3f/9+r9vfcsstMBqNAGonfOTl5eGLL74AAAwfPlx9zk/nuap/DJxKhw4d1Odh69atmD9/Ps4++2z19nfffdfnfU2fPh0zZ87EDz/8AJvN5jFxBvjjPcO17q7PUUREBOLi4jzu509c9enTB8888wzatGmDbdu2Yc6cObjxxhvRu3dvdOvWDb/88ovPbaKTY9LUAvTs2RNfffUVYmNjAQDffPMNhg8fjtLS0gbvM3v2bDgcDqxYsQJWqxU2mw333HMPVqxYoW5T/wP5tddeQ9++fdG3b1/84x//UMu//fZbdcaJ3W53m53SoUMHjB07Fu+88w6GDBmilrt+GJ2Ma2L0/PPPq2tORURE4Nprr1Vvc51N99xzz6G8vBwigj/96U8n3b/VavWpHgDw1ltvqb/fd9996pvgjBkzTnnf0NBQ9feG1jBqzNpGy5cvR0lJifr3mDFj1Nfop59+UsvrJ7a+SkpKUn/ftGmTR6IiIjh69GiT76Ox6r++rnU566yzvNZFRPDEE08AAOLj49XtDx06pP5+8OBBr4/nOrPK9UvI3r17Pbb1ty6uTnU8xcXFISQkBABQUlKC7Oxsr/WtM3PmTADAiRMnMG7cOPVL1x133HHS+wHA5ZdfDpPJBAAoKirCxx9/7Ha7iGDJkiXq31dccYX6e7t27XDllVcCAFauXIkXXnhBTfZuu+02dTvX5+r222/3+jw5nU5kZGR41M+fGC8uLlZ/NxqN6NevH6ZNm4Yvv/xSLff1vQtwf1/68MMPYbfbISJuXzrrJCQkqL+7HmslJSVeH9PfuLr33nuRl5eHzMxMLF++HA8//DCMRiP27t2rvv50+pg0tRDp6en4+uuv1Tf5devWYdiwYbDZbA3eJzQ0FFdddRUef/xxtWzq1Klq0lP/A7kh4jK1/dChQ0hNTcVDDz2EjRs3wmazwW63Y8OGDdi+fbt6H9dvbiczZswYREZGqm1yOBwAgGuvvdbtG3bdBwRQm1ApioJPPvlETbICwfUxrFYrQkNDsW7duoBN63f9Nrl9+3av30rr8zUZck1s/TF69Gj197vuugtbt26F3W5HQUEBVq5cieuuuw5z585t8n0ESpcuXdQeuF27dmHmzJnIyclBVVUV9u/fj4ULF6J3795qUuS6ZMecOXOQlZWFvLw83H///V7379rb9d///hc1NTU4evQonn322dOuiz/CwsIwYsQI9e8///nPyMzMREVFBXbv3u32pQeojbO0tDQAtV+6AOCMM87A0KFDT/lYCQkJuPvuu9W/77rrLqxcuRJ2ux1HjhzBpEmT8OuvvwIAoqOjPb5k1CVHDodDfS/q0KGDW/2HDx8Os9kMoLY37PXXX0dxcTEqKiqQmZmJ//u//8PAgQN9e3JO4plnnsFFF12El156CXv27IHD4UBRUZFbnNV/76qL2/z8fBw5csTtNtf3jMjISFRXV2Px4sXYtm2bx2O7HmvPPfccfvnlFxQVFWH69Oleeyr9iasdO3bg//7v//DDDz+gbdu2uPrqq3Httdeqz+mpkmryQ9Oc9aPT1dDilj///LPbQONzzz1XHWjY0Ow5h8MhXbt2VW97+umnRcR9sUTXc+x1XGe/1E1t92WcxvXXX+9XW13H7NT91J8R5jquou7HYDDIGWec4XWMSV1ZQwvSeRtf9O6773ptT7du3bw+rw1NE3Z9HVxn8rjOMDpV/UTcl4IwmUweSx6IiIwdO1bdV92AXn9nz11yySUnfT1P1WZ/99EQ1+19nT3nzebNmyUqKuqk9anbf0Oz59q3b+/1NSouLnabrRUeHi4Gg0HCw8O9xqs/dTnZsgXe6nLgwIFTzp5z5TqgHID885//PMmr4e50FresqanxqOcjjzzisd2LL77Y4Oy5+m1v6Bg/lYcffvikbTAYDG5j20TcB9zXf31cZ/rW/VitVnUguOt7gMPhkHPPPddj+6ioKLeZb3X8iat169addLv647So8Zg0NVMNJU0iIjt37pTk5GT19n79+klBQUGDSZOI+0DNNm3aSHZ2ttsHsrcZJ/Xf7NasWSPl5eWyePFidW2S+Ph4CQkJkaioKDnvvPNk/vz5btPyfVE/4L2tPVRdXS1z5syR1NRUMZvNcs4558gnn3zi1ubTTZpERBYtWiTdunUTs9ksZ555prz00ktur8XpJE0VFRVy1113Sfv27dXn/mRJk+sg2jFjxnjdZtWqVR7Pm7/rNFVVVcnzzz8vAwYMkKioKDGZTNKhQwe5+OKL5R//+Ic6gzFQ+2iIt0TCG19WJT9w4IDceeed0qVLFzGbzRIRESFdu3aV6667TpYuXarO5BL5Y52miIgIiYyMlD/96U9y4MCBBo+hLVu2yEUXXSQRERGSmJgot99+u3zwwQcNxquvdfE3aRIRKSgokIceekh69uwpFotFXafJdap7nbKyMnXws8lkktzc3Aafv4Z88sknMnLkSElOTpbQ0FCJiIiQPn36yKxZs9xmitb36KOPuiUm3iZniIisX79err32WklOTpaQkBCJjY2VXr16yZQpU9yuftDYpGnXrl0yZ84cGTJkiKSlpUlERISEhIRIcnKyjBo1ym3pkzoHDhyQq6++2m3WXd3rU1ZWJtOmTZN27dpJWFiYDBgwQDZs2OD2/uL6HlA3Wy4mJkYsFosMGTJEtm3bpiZN8fHxbo/ta1wdOXJE7r77bunXr586icVqtUqfPn3kySef9FgvihpPEfHhHAERkc7UjSVKSUnR7PqETenw4cM466yzUFZWhvHjx7fa1eSbs02bNqFDhw7qAHi73Y7HH38cf//73wEAN954o9s4KWp+mDQREXnRWpKmjz76CA888AAOHTqEyspKWCwWbN++HV27dg121XRn4sSJeO211xAdHY2oqCjk5uaqY0w7deqE9evX+zzbl4KDA8GJiFqx4uJi7NmzR53VtWLFCiZMQTJs2DAMGjQIoaGhyMnJQWhoKPr27Yu//e1v2L59OxOmFoA9TUREREQ+CGpP09q1a3H11VejXbt2UBTFY/0Pb7799lukp6cjLCwMnTt3xqJFi5q+okRERKR7QU2aysrKcM455+Df//63T9tnZWVhxIgRGDRoELZt24aHHnoI99xzD5YvX97ENSUiIiK9azan5xRFwUcffYRRo0Y1uM0DDzyATz/9FDt37lTLpkyZgu3bt/t02Q4iIiKixgo59SbNx6ZNm9xWVQWAjIwMvPLKK6iqqnK79ECd+pf9cDqdKCwsRFxcXKMua0FERESBJSIoKSlBu3btYDA03zlqLSppOnbsmNv1eIDa6/NUV1cjPz8fycnJHveZO3cuHnvsMa2qSERERI106NAhdOjQIdjVaFCLSpoAz4tX1p1dbKjXaNasWW7XQiouLkanTp1w8OBBREVFqfdVFEW9CKLrY52s3NvVvquqqhASEuJWn4a2NxgMHvv2t7yxdfe1Tf6Ws01sk5ZtEhE4HA6YTCYYjcZW0abG1p1tYpu0aJOiKLDb7QgNDXX7nDvdNtlsNqSkpKjXIm2uWlTS1LZtWxw7dsytLDc3FyEhIW4XQ3VlNpvVixa6atOmjZo0ERERUfDUnZJr7sNmmu+JQy8GDBiA1atXu5WtWrUK/fv39zqeSWtOpxPHjx/32gNFRIHHmCPSlt5jLqhJU2lpKTIzM5GZmQmgdkmBzMxMZGdnA6g9tTZ+/Hh1+ylTpuDgwYOYMWMGdu7ciSVLluCVV17BzJkzg1F9r5rJZEQi3WDMEWlLzzEX1NNzW7ZswaWXXqr+XTf2aMKECVi6dClycnLUBAoA0tLSsHLlSkyfPh0vvPAC2rVrh3/9618YM2aM5nUnIiIifWk26zRpxWazITo6GsXFxQEf0+R0OpGbm4vExMRmPWWSqLVgzBFpq6lirik/mwOJ7zIBpCgK138i0hBjjkhbeo85Jk0BpCgKjEajbg8mIq0x5oi0pfeYY9IUQHXdlnqdVUCkNcYckbb0HnNMmoiIiIh8wKSJiIiIyAdMmoiIiIh8wKQpgAwGA6c+E2mIMUekLb3HnD5b3UREBDU1NbpeLZVIS4w5Im3pPeaYNAWQiKCgoEC3BxOR1hhzRNrSe8wxaSIiIiLyAZMmIiIiIh8waQowva6SShQsjDkibek55kKCXYHWxGAwICkpKdjVININxhyRtvQec+xpCiARgd1u1+0AOSKtMeaItKX3mGPSFEAigqKiIt0eTERaY8wRaUvvMcekiYiIiMgHTJqIiIiIfMCkKcBCQji2nkhLjDkibek55vTb8iZgMBgQHx8f7GoQ6QZjjkhbeo859jQFkIigvLxctwPkiLTGmCPSlt5jjklTAIkIbDabbg8mIq0x5oi0pfeYY9JERERE5AMmTUREREQ+YNIUQIqiwGQy6fq6PERaYswRaUvvMcfZcwGkKApiY2ODXQ0i3WDMEWlL7zHHnqYAEhGUlJTodoAckdYYc0Ta0nvMMWkKIBFBWVmZbg8mIq0x5oi0pfeYY9JERERE5AMmTUREREQ+YNIUQIqiwGKx6HZWAZHWGHNE2tJ7zHH2XAApioLo6OhgV4NINxhzRNrSe8yxpymARATFxcW6HSBHpDXGHJG29B5zTJoCSERQUVGh24OJSGuMOSJt6T3mmDQRERER+YBJExEREZEPmDQFkKIoCA8P1+2sAiKtMeaItKX3mOPsuQBSFAWRkZHBrgaRbjDmiLSl95hjT1MAiQgKCwt1O0COSGuMOSJt6T3mmDQFkIjA4XDo9mAi0hpjjkhbeo85Jk1EREREPmDSREREROQDJk0BpCgKoqKidDurgEhrjDkibek95jh7LoAURYHVag12NYh0gzFHpC29xxx7mgLI6XQiPz8fTqcz2FUh0gXGHJG29B5zTJoCrLq6OthVINIVxhyRtvQcc0yaiIiIiHzApImIiIjIB0yaAkhRFMTExOh2VgGR1hhzRNrSe8xx9lwAKYoCs9kc7GoQ6QZjjkhbeo859jQFkNPpxPHjx3U7q4BIa4w5Im3pPeaCnjQtXLgQaWlpCAsLQ3p6OtatW3fS7d966y2cc845sFqtSE5Oxl/+8hcUFBRoVNtT0+v1eIiChTFHpC09x1xQk6b33nsP06ZNw8MPP4xt27Zh0KBBGD58OLKzs71uv379eowfPx6TJk3Cr7/+imXLluGHH37ArbfeqnHNiYiISG+CmjQ9++yzmDRpEm699VZ0794dCxYsQMeOHfHiiy963f67775Damoq7rnnHqSlpeGiiy7C7bffji1btmhccyIiItKboCVNDocDW7duxbBhw9zKhw0bho0bN3q9z8CBA3H48GGsXLkSIoLjx4/jgw8+wJVXXqlFlU9JURTExcXpdlYBkdYYc0Ta0nvMBW32XH5+PmpqapCUlORWnpSUhGPHjnm9z8CBA/HWW29h7NixqKysRHV1Na655ho8//zzDT6O3W6H3W5X/7bZbABqB7PVDWRTFAWKokBE3M7Vnqq8/kA4RVFgNBp93t5gMHhs6295Y+vuT5v8KWeb2Cat21S3T2/7bqltakzd2Sa2Sas2ebvtdNvUUgaWB33JgfrZat2bnzc7duzAPffcg0ceeQQZGRnIycnBfffdhylTpuCVV17xep+5c+fiscce8yjPy8tDZWUlAMBisSA6Oho2mw0VFRXqNuHh4YiMjERRUREcDodaHhUVBavVisLCQrfl5KOjo1FcXOzRhri4OBiNRuTm5rrVITExETU1NW4D2RVFQVJSEhwOB4qKitTykJAQxMfHo6KiQk38AMBkMiE2NhalpaUoKytTywPVppiYGJjNZuTl5bkd9GwT29Qc2uR0OlFcXIzo6Gi0bdu2VbSpNb5ObFPraVNCQgKOHj0Ko9EIg8EQsDaVlJSgJVAkSMPgHQ4HrFYrli1bhtGjR6vlU6dORWZmJr799luP+9x8882orKzEsmXL1LL169dj0KBBOHr0KJKTkz3u462nqWPHjigqKkJUVBSAwGXyIoK8vDzEx8erB9PJttfbtxO2iW0KdJucTify8vKQkJCAkJCQVtGmxtadbWKbtGgTABw/fhwJCQlun3On2yabzYaYmBgUFxern83NUdB6mkwmE9LT07F69Wq3pGn16tUYOXKk1/uUl5cjJMS9ykajEUDDUyDNZrPXhbjquhhd1b2Q9TVUXv/+dQedt317274xj9nU5d7q6G8528Q2Naa8sW1SFMXt99bQpuZQd7aJbfK2b6fTqcbc6X6GupY3VK/mJqi1nDFjBl5++WUsWbIEO3fuxPTp05GdnY0pU6YAAGbNmoXx48er21999dX48MMP8eKLL2L//v3YsGED7rnnHpx33nlo165dsJpBREREOhDUMU1jx45FQUEB5syZg5ycHPTs2RMrV65ESkoKACAnJ8dtzaaJEyeipKQE//73v3HvvfeiTZs2uOyyy/DPf/4zWE1wYzAYkJiY2GIyZqKWjjFHpC29x1zQxjQFi81mUwdsB/q8qYiguroaISEhXrsiiSiwGHNE2mqqmGvKz+ZA0meq2EREBAUFBQ2OryKiwGLMEWlL7zHHpImIiIjIB0yaiIiIiHzApCnAOK6CSFuMOSJt6Tnmgr4ieGtiMBg8LgtDRE2HMUekLb3HHHuaAkhEYLfbdTtAjkhrjDkibek95pg0BZCIoKioSLcHE5HWGHNE2tJ7zDFpIiIiIvIBkyYiIiIiHzBpCrD6FxQmoqbFmCPSlp5jTr8tbwIGgwHx8fHBrgaRbjDmiLSl95hjT1MAiQjKy8t1O0COSGuMOSJt6T3mmDQFkIjAZrPp9mAi0hpjjkhbeo85Jk1EREREPmDSREREROQDJk0BpCgKTCaTrq/LQ6QlxhyRtvQec5w9F0CKoiA2NjbY1SDSDcYckbb0HnPsaQogEUFJSYluB8gRaY0xR6Qtvccck6YAEhGUlZXp9mAi0hpjjkhbeo85Jk1EREREPmDSREREROQDJk0BpCgKLBaLbmcVEGmNMUekLb3HHGfPBZCiKIiOjg52NYh0gzFHpC29xxx7mgJIRFBcXKzbAXJEWmPMEWlL7zHHpCmARAQVFRW6PZiItMaYI9KW3mOOSRMRERGRD5g0EREREfmASVMAKYqC8PBw3c4qINIaY45IW3qPOc6eCyBFURAZGRnsahDpBmOOSFt6jzn2NAWQiKCwsFC3A+SItMaYI9KW3mPO76Rp4sSJWLt2bVPUpcUTETgcDt0eTERaY8wRaUvvMed30lRSUoJhw4aha9eueOKJJ3DkyJGmqBcRERFRs+J30rR8+XIcOXIEd999N5YtW4bU1FQMHz4cH3zwAaqqqpqijkRERERB16gxTXFxcZg6dSq2bduGzZs3o0uXLrj55pvRrl07TJ8+HXv27Al0PVsERVEQFRWl21kFRFpjzBFpS+8xd1oDwXNycrBq1SqsWrUKRqMRI0aMwK+//ooePXpg/vz5gapji6EoCqxWq24PJiKtMeaItKX3mPM7aaqqqsLy5ctx1VVXISUlBcuWLcP06dORk5OD1157DatWrcIbb7yBOXPmNEV9mzWn04n8/Hw4nc5gV4VIFxhzRNrSe8z5vU5TcnIynE4n/vznP2Pz5s3o06ePxzYZGRlo06ZNAKrX8lRXVwe7CkS6wpgj0paeY87vpGn+/Pm47rrrEBYW1uA2MTExyMrKOq2KERERETUnfp+eu+aaa1BeXu5RXlhYCJvNFpBKERERETU3fidN48aNw7vvvutR/v7772PcuHEBqVRLpSgKYmJidDtAjkhrjDkibek95vxOmr7//ntceumlHuWXXHIJvv/++4BUqqVSFAVms1m3BxOR1hhzRNrSe8z5nTTZ7Xavg8CqqqpQUVERkEq1VE6nE8ePH9ftrAIirTHmiLSl95jzO2k699xzsXjxYo/yRYsWIT09PSCVasn0ej0eomBhzBFpS88x5/fsuccffxxDhgzB9u3bcfnllwMA/ve//+GHH37AqlWrAl5BIiIioubA756mCy+8EJs2bULHjh3x/vvvY8WKFejSpQt++uknDBo0qCnqSERERBR0iuisn81msyE6OhrFxcWIiooK6L5FBNXV1QgJCdHtIDkiLTHmiLTVVDHXlJ/NgeT36TmgdiDY3r17kZub6zEY7OKLLw5IxVoiRVFgNBr55k2kEcYckbb0HnN+J03fffcdbrjhBhw8eNBjMJiiKKipqQlY5Voap9OJ3NxcJCYmwmA4rWshE5EPGHNE2tJ7zPmdNE2ZMgX9+/fH559/juTkZN1mm0RERKQvfidNe/bswQcffIAuXbo0RX2IiIiImiW/+9bOP/987N27N2AVWLhwIdLS0hAWFob09HSsW7fupNvb7XY8/PDDSElJgdlsxhlnnIElS5YErD5ERERE3vjd0/TXv/4V9957L44dO4ZevXohNDTU7fbevXv7vK/33nsP06ZNw8KFC3HhhRfiP//5D4YPH44dO3agU6dOXu9z/fXX4/jx43jllVfQpUsX5Obmel2hPBgMBoNuz/MSBQNjjkhbeo85v5cc8PZEKYoCEfF7IPj555+Pfv364cUXX1TLunfvjlGjRmHu3Lke23/xxRcYN24c9u/fj9jYWH+qreKSA0StB2OOSFt6X3LA71QxKyvL42f//v3q/75yOBzYunUrhg0b5lY+bNgwbNy40et9Pv30U/Tv3x9PPfUU2rdvj27dumHmzJnN5pp3IoKCggJdLzFPpCXGHJG29B5zfp+eS0lJCcgD5+fno6amBklJSW7lSUlJOHbsmNf77N+/H+vXr0dYWBg++ugj5Ofn484770RhYWGD45rsdjvsdrv6t81mA1A7bbJujSlFUdTeMtcD4VTl9deoqtumfnlD2xsMBo99+1ve2Lr72iZ/y9kmtknLNjmdTogInE5nq2lTY+vONrFNWrQJgBpzgWxTS7kAcKMWt3zjjTewaNEiZGVlYdOmTUhJScGCBQuQlpaGkSNH+rWv+t17daf5vHE6nVAUBW+99Raio6MBAM8++yyuvfZavPDCC7BYLB73mTt3Lh577DGP8ry8PFRWVgIALBYLoqOjYbPZ3HqtwsPDERkZiaKiIjgcDrU8KioKVqsVhYWFbuOp6uqUl5fn1oa4uDgYjUbk5ua61SExMRE1NTUoKChwez6SkpLgcDhQVFSkloeEhCA+Ph4VFRVq4gcAJpMJsbGxKC0tRVlZmVoeqDbFxMTAbDYjLy/P7aBnm9im5tAmp9OJ4uJiiAjatm3bKtrUGl8ntqn1tCkhIQHV1dXIzc1Vh+sEok0lJSVoCfwe0/Tiiy/ikUcewbRp0/D444/jl19+QefOnbF06VK89tpr+Oabb3zaj8PhgNVqxbJlyzB69Gi1fOrUqcjMzMS3337rcZ8JEyZgw4YNbrP3du7ciR49emD37t3o2rWrx3289TR17NgRRUVF6nnTQPY05efnIy4uzm3sF7+dsE1sU9P1NOXn5yM+Ph4hISGtok2NrTvbxDZp0SYAyM3NRXx8vNvn3Om2yWazISYmptmPafK7p+n555/HSy+9hFGjRuHJJ59Uy/v374+ZM2f6vB+TyYT09HSsXr3aLWlavXp1g71VF154IZYtW4bS0lJEREQAAHbv3g2DwYAOHTp4vY/ZbIbZbPYoNxgMHoPa617I+hoq9zYovv7pxlNt7+9jNnV5QzMi/Clnm9imxpQ3pk0GgwFt27YNet35OrFNemqTa8ydTh1dy1vKbLxGDQTv27evR7nZbHbrfvPFjBkz8PLLL2PJkiXYuXMnpk+fjuzsbEyZMgUAMGvWLIwfP17d/oYbbkBcXBz+8pe/YMeOHVi7di3uu+8+3HLLLV5PzWlNRGC3271m5kQUeIw5Im3pPeb8TprS0tKQmZnpUf7f//4XPXr08GtfY8eOxYIFCzBnzhz06dMHa9euxcqVK9XB5jk5OcjOzla3j4iIwOrVq3HixAn0798fN954I66++mr861//8rcZTUJEUFRUpNuDiUhrjDkibek95vw+PXfffffhrrvuQmVlJUQEmzdvxjvvvIO5c+fi5Zdf9rsCd955J+68806vty1dutSj7KyzzsLq1av9fhwiIiKi0+F30vSXv/wF1dXVuP/++1FeXo4bbrgB7du3x3PPPYdx48Y1RR2JiIiIgq5RSw5MnjwZkydPRn5+PpxOJxITEwNdrxYrJKRRTykRNRJjjkhbeo6502p5fHx8oOrRKhgMBj4nRBpizBFpS+8x53fSlJaW5nXqYB1/LqXS2ogIKioqYLFYTvocEVFgMOaItKX3mPM7aZo2bZrb31VVVdi2bRu++OIL3HfffYGqV4skIrDZbAgLC9PlwUSkNcYckbb0HnN+J01Tp071Wv7CCy9gy5Ytp10hIiIiouYoYEtwDh8+HMuXLw/U7oiIiIialYAlTR988AFiY2MDtbsWSVEUmEwmXXZZEgUDY45IW3qPOb9Pz/Xt29ftyRIRHDt2DHl5eVi4cGFAK9fSKIqi+8SRSEuMOSJt6T3m/E6aRo0a5fa3wWBAQkICLrnkEpx11lmBqleLJCLqxYT1moUTaYkxR6Qtvcec30nT7Nmzm6IerYKIoKysDOHh4bo8mIi0xpgj0pbeY87vpMlms/m8bVRUlL+7JyIiImqW/E6a2rRpc8rsUkSgKApqamoaXTEiIiKi5sTvpOnVV1/Fgw8+iIkTJ2LAgAEAgE2bNuG1117D3LlzkZqaGug6thiKouh2lVSiYGDMEWlL7zGniIj4c4fLL78ct956K/785z+7lb/99ttYvHgx1qxZE8j6BZzNZkN0dDSKi4t5+pCIiKgZaCmfzX6v07Rp0yb079/fo7x///7YvHlzQCrVUokIiouL4WceSkSNxJgj0pbeY87vpKljx45YtGiRR/l//vMfdOzYMSCVaqnqLmSo14OJSGuMOSJt6T3m/B7TNH/+fIwZMwZffvklLrjgAgDAd999h3379vEyKkRERNRq+d3TNGLECOzevRvXXHMNCgsLUVBQgJEjR2L37t0YMWJEU9SRiIiIKOj87mkCak/RPfHEE4GuS4unKIpuF/wiCgbGHJG29B5zjbpg77p163DTTTdh4MCBOHLkCADgjTfewPr16wNauZZGURRERkbq9mAi0hpjjkhbeo85v5Om5cuXIyMjAxaLBT/++CPsdjsAoKSkRPe9TyKCwsJC3Q6QI9IaY45IW3qPOb+Tpn/84x9YtGgRXnrpJYSGhqrlAwcOxI8//hjQyrU0IgKHw6Hbg4lIa4w5Im3pPeb8Tpp+++03XHzxxR7lUVFROHHiRCDqRERERNTs+J00JScnY+/evR7l69evR+fOnQNSKSIiIqLmxu+k6fbbb8fUqVPx/fffQ1EUHD16FG+99RZmzpyJO++8synq2GIoioKoqCjdDpAj0hpjjkhbeo85v5ccuP/++1FcXIxLL70UlZWVuPjii2E2mzFz5kzcfffdTVHHFkNRFFit1mBXg0g3GHNE2tJ7zPl1wd6amhqsX78evXr1QlhYGHbs2AGn04kePXogIiKiKesZME15UUCn04nCwkLExsbCYGjUag5E5IPKqhqs/DkHX/56DPnF5YiPtiLj7LYY0SsZYaHGYFePqNVqqs+5lnLBXr+SJgAICwvDzp07kZaW1lR1alJNnTTl5uYiMTGRSRNRE1m94zjuXZYJW0U1DArgFKj/R1lC8Ox1fTCkR1Kwq0nUKjXV51xLSZr8bnGvXr2wf//+pqgLEdFJrd5xHLe9sQUlFdUAahMl1/9LKqox+Y0tWL3jeJBqSEStmd9J0+OPP46ZM2fis88+Q05ODmw2m9sPEVFTqKyqwb3LMgEBGuoel9//mbksE5VVNdpVjoh0we+B4FdccQUA4JprrnEbPS8iUBQFNTX6faNSFAUxMTG6nVVA1JRW/pwD2+89TCcjAIorqvHfX3Iwum+Hpq8YkY7o/XPO76Tpm2++aYp6tAqKosBsNge7GkQtktMpOFFRhYJSO/JLHSgos6Og1FH7d5kDX/lxyk0B8EnmUYw8pz0MBn2+uRM1Bb1/zvk8EHz8+PF44YUXEBkZCQDYvn07evTo4XYplZagqQeC5+XlISEhgQPBiQBUOGqQX2pHQdnvyU9dQuSSFNXdXljmQI0zsJdmsIQakRJnReeEcKTGhSMtPlz9PTbcpNtvy0SN1VSfcy1lILjPSZPRaEROTg4SExMB1F42JTMzs8WtAs7Zc0SNV+MUFJbVS3jcEqC6JKj273JH8z1dHxUWgrT48N9/IpAab0Xn3/+PDGtZXwaJtKL32XM+n56rn1vp9WJ9RK2JiKDMUfPHKbHfe33yS37/v15SVFjuQKBD32Q0IC7ChPgIM+IiTIgLNyM+wqT+XnfbD1mFeOyzHT7vt0dyFCqrapBdWI5qLz1YtspqbD9cjO2Hiz1ui48wo3N8OFLjrUiLj0Da7/+nxFm5DhSRjvk9pomImreqGieKyhzI89ILpCZFpX/0FNmrnQGvQ4w1FHERZsSF/5EMeSZFtX9HmkN8Ok3WJTEC8/+3GyUV1Q3OngNqxzNFWULw4Z0DERZqRFWNE0eKKpCVX4b9+WU4kF+GrN9/jhZXeE0C838/lbj5QKH7vhWgXbQFaV4Sqg4xFoQa2cNM1Jr5lTTt2LEDx44dA1D7DXXXrl0oLS1126Z3796Bq10LoygK4uLiOE6CAkpEYKusVhOeul4h114g14ToRHlVwOtgDjEgPsKM+Egz4sN/7wWqlxTFhZsRH2lCjNXUJMlDWKgRz17XB5Pf2AKlgWUHlN//mXddH7VHKNRoQGp8OFLjw3Fpve0rq2pwsKBcTaLqEqr9+WXIL7V77F8EOHKiAkdOVGB9veuWhxgUdIy11iZUceFISwj/vbcqHMlRYRyQTq2C3j/nfB7TZDAYoCiK19NydeUtYcmBpj5v6nQ6OZ6JTsleXYPCMgfySxzId5klVv+UWH5J7f9VNYE9J2ZQgNjwP05/1SVACZG1/9f1AsX/frvVZGw2b5KrdxzHzGWZKPayIni0JQTzArQieEllFQ7klyOroAxZeWXIyi9FVkE5svJKYas89dIHrswhBnUgelpCONJ+T6pS48IRH8EB6dSyNMXnXEsZ0+Rz0nTw4EGfdpiSknJaFWpqHAhOTcHpFBRXVLn0+vzRA1SbBNUlQrV/l/j5oeuLcJMRcRF/nPqKj3BPiuLDTYj/PSlqYzXB2IJ7PiqravDfX3LwxS/HkFdchoTocFzRsy2G92z6a8+JCIrKq5CVX4r9eWU4UFB3uq8cWfmlqKzy73RnpDlETaD+GJhe20MVbeGAdGpe9D4Q3O9rz7V0TJrIV5VVNW5jfwpKvfUK1d7WFNPljQZF7fWpTYDqfq8bI+SSFIWbYTHpb4Byc4s5p1NwvKTS43RfVn4ZsgvL/e4xjAs3qQlUWvwfp/tS48J1+XpT8Ok9aeJAcNKNGqegqNzhtmBigUtSVH9BxbImmC4fGRZSOzaofi+Q20yx2nFCUWGhHAfTwhgMCpKjLUiOtmDgGfFut1XXOHHkRIVbQrX/99+PnPA+IL2grLZ3csvBIo/bkqPD3Hqm6pKrjjFWmEKCn0AStUZMmqjFEhGUO2rUHqC6afJ/rCjtcDktVtsbFODOIHW6fP3p8d5mjcWGm2AOYe+AXoUYDUiJC0dKXDguOdP9tsqqGhwq/GNAuutPbonngHQAyCmuRE5xJTbuK3ArNxoUdIyxqL1TakIVF452bSwt+rQsUbDx9FyAcSD46ambLp9f6rlitOtaQnW3+zt+xBdtrKFq0uM6M0w9JVaXFEWafZ4uT02ntcdcqb1aPc2nnu4rKMP+vDIUV/g3U9IUYkBqnFWd3ZfmMjg9IcLMY5l8oueB4OxpCiARQU1NDRRF4ZvP70QEJfbqP06JeV1F+o+kqKgpp8u7JDx1p8Tqrx0UE9400+Wpaegh5iLMIejZPho920d73FZU5lBn9x0o+P103++/e1uN3VHtxO7jpdh9vNTjtnCTUR2Q3jk+3OX3CERbOSCdaukh5k6mUUlTdXU11qxZg3379uGGG25AZGQkjh49iqioKERERAS6ji2GiKCgoACJiYmt+mByVDt9voxGQakDjprA9gYpChBrNbmsDVR3Osx7UtScpstTYOkl5hoSE16b6PfrFONWLiLILbF7Pd2XXVDuNSbLHDX45YgNvxyxeT6ONVS93EzdYp61i3uGw2rid2890XvM+X20Hzx4EFdccQWys7Nht9sxdOhQREZG4qmnnkJlZSUWLVrUFPVs1iqrarDy5xx8+esx5J0oQ0Kbw8g4uy1G9Gr66c+BIFI7Xd51gcSCUjvySt3HBNUlSf6uUeOLuuny3i6j8ceCirW3x7Tw6fJETU1RFCRFhSEpKgwXdI5zu63GKTh6osJjdfSs/DIcLir3Ou6vqLwKRdkn8GP2CY/b2kaFeayOnhYfjk6xHJBOrY/fSdPUqVPRv39/bN++HXFxfwTj6NGjceuttwa0ci3B6h3Hce+yTNhcF9o7Woovfz2OR1f8imcDtNCevyqralyuIeaydpDLAop1SVJhmcPrtblOh9Gg/L54YsMLJrquKM3p00TaMP6+cnnHWCsGd0twu81eXYNDhRVus/vqEqtjtkqv+ztmq8QxWyW+2+9+yRmDAnSIsSL196US6mb3dY7ngHRqufxOmtavX48NGzbAZDK5laekpODIkSMBq1hLsHrHcdz2xhb1eg7Oev+XVFRj8htbsPjm/hh6molTjVNwotzhvmK0y4Bo914iB0rtge8NqpsuH+dyGY34BtYOirZwujxpQ4+nCJqKOcSILokR6JLoOcyi3FFdu0J6/u/jp35fJf1AQTkKyxwe2zsFyC4sR3ZhOdbuznO7zWQ0oFOc1WN2X+eEcCRGckB6c6fn18fvpMnpdHq9VMrhw4cRGRkZkEq1BJVVNbh3WSbQwDWwgNpyRYCZyzLx/UNDPE7VlTtqB0jnlTZ8GY26U2JNMV0+1Kh4rBUUH9FQrxCny1PzYzAYkJSkfU+uHllNIejRLgo92nnObCour6odkJ5fWnvJmYLa1dEP5Jd7/QLnqHFib24p9uZ6Dki3mowes/vqeqhiwk0e25O29B5zfidNQ4cOxYIFC7B48WIAtRlnaWkpZs+ejREjRgS8gs3Vyp9zYKs4dW+OACiuqMaEJd8jwhzqtqBiRVXgF0+smy7v7TIaCfUGSkeFcbo8tWwiAofDAZOJ128LpmhrKPpY26BPxzZu5SKCvFK72+w+dfmEgnI4qj0HpJc7arAjx4YdOZ4D0ttYQ9XZffXXoQo3c0C6FvQec36v03T06FFceumlMBqN2LNnD/r37489e/YgPj4ea9euRWJiol8VWLhwIZ5++mnk5OTg7LPPxoIFCzBo0KBT3m/Dhg0YPHgwevbsiczMTJ8fL1BrQUx5YytW7TgW8N6f+kwhBiSoA6Trpsi7D5SuW0AxltPlSWea22VUyHc1TkFOcYXH6ugH8stwqKjC78sSJUaavY6f6hRnZS95APEyKn5q164dMjMz8c477+DHH3+E0+nEpEmTcOONN8Jisfi1r/feew/Tpk3DwoULceGFF+I///kPhg8fjh07dqBTp04N3q+4uBjjx4/H5ZdfjuPHj/vbhIA4Ud6402X1p8vH1Tsl5nax1QgzwjldnohaIaNBQYcYKzrEWDGoq/uAdEe1E4eLPFdIP5BfhqPF3gek55bYkVtix+Ys9wHpigK0b2PxuNxM5/hwtG9jQQi/aJIfgroi+Pnnn49+/frhxRdfVMu6d++OUaNGYe7cuQ3eb9y4cejatSuMRiM+/vjjZt/TpAAY2CUOC8b2RYw1lEFKFCDsadKfCkcNDhR4zu7Lyi9DgZcB6ScTaqydSdj594HoaQl/JFZto8L4hdUL9jT54NNPP/V5h9dcc41P2zkcDmzduhUPPvigW/mwYcOwcePGBu/36quvYt++fXjzzTfxj3/845SPY7fbYbf/ce0mm632PLnT6YTTWXs+vW5lUxGBaw55svJhZyfhi1+P+dRWAXBtegfEhYeqjw3UDqirv29/yxtT95OV19WtseVsE9ukZZucTieMRqN6WYfW0KbG1l0vbTKHKDirbSS6J0d5bF9SWY0Dvw9Cr53dVzt26kB+GUq8DEivqhHsz6udCVifJdSIFJcZfqlxVnROqF2Dqo3FfTyonl4nAGrMBbJN9ffXXPmUNI0aNcqnnSmK4nVmnTf5+fmoqanxGIWflJSEY8e8JyN79uzBgw8+iHXr1iEkxLczi3PnzsVjjz3mUZ6Xl4fKytpuXovFgujoaNhsNlRUVKjbhIeHIzIyEkVFRXA4/vgGExUVhRG9kjH7k19Qaq9pcPYcUNvLFBkWgit6JCE3N9fttsTERNTU1KCg4I8LbiqKgqSkJDgcDhQV/XFl85CQEMTHx6OiokJN/ADAZDIhNjYWpaWlKCv7I/Ab0yar1YrCwkJUV//x5hITEwOz2Yy8vDy3gz4uLg5Go5FtYpuaRZvy8/NbXZuA1vc6adGmczq2QWqUgguSQwDUXnrGarXCYTBj+/4c7M8rxaEiO7JPVOKorQoHCytg9zIgvaKqBruOlWDXsRKP2yLNRnRsY0bHmDB0amPG2SmJSImxIEqpRLjZ2Opfp6ioKOTn5we0TSUlns9zcxS003NHjx5F+/btsXHjRgwYMEAtf/zxx/HGG29g165dbtvX1NTgggsuwKRJkzBlyhQAwKOPPnrK03Peepo6duyIoqIitQuwsZn86l9zcNubPza47IDy+z+Lb07H0B5t+S2SbWKbAtwmEUFFRQUsFguMRmOraFNj6842Na5cBDhyotxtVt+B/NplEw4Vlvu98G98hMll3anaVdJTYq1IibOqy8605NdJURSUl5cjLMz99OXptslmsyEmJqbZn547raSpsrISYWFhjbqvw+GA1WrFsmXLMHr0aLV86tSpyMzMxLfffuu2/YkTJxATEwOj8Y9ZEHVvmkajEatWrcJll112yscN9HnT1TuOY+ayTBS7rgj++//RlhDMC9KK4ER6wDFN1JSqapw4XFThMruvVF3g82hxBfz59FQUoF20xWMwemp8ODrEWFrMzGeOafJTTU0NnnjiCSxatAjHjx/H7t270blzZ/ztb39DamoqJk2a5NN+TCYT0tPTsXr1arekafXq1Rg5cqTH9lFRUfj555/dyhYuXIivv/4aH3zwAdLS0vxtSkAM7ZGE7x8agv/+koMvfjmGvOIyJESH44qebTG8Z8u49hwREXkKNRrUJOfSerdVVtXgYIHrDL/ahGp/fhnyS+0e+6rt0arAkRMVWL833+22EIOCTrFWde0p16UT2kaF8eoKzYjfSdPjjz+O1157DU899RQmT56slvfq1Qvz58/3OWkCgBkzZuDmm29G//79MWDAACxevBjZ2dnq6bdZs2bhyJEjeP3112EwGNCzZ0+3+ycmJiIsLMyjXGthoUaM7tsBI89px2+9REQ6EBZqxJltI3FmW88rYZRUVv2eQNX1TJXWrpKeV+r1gufVTsH+33uz6jOHGNTTfa6z+9LiwxEXrs8FJoPJ76Tp9ddfx+LFi3H55ZeryQ0A9O7d22Mc0qmMHTsWBQUFmDNnDnJyctCzZ0+sXLkSKSkpAICcnBxkZ2f7W8WgURRFt6ukEgUDY46ao8iwUPTqEI1eHaLdykUEhWUO9dp9Bwpqe6nqfq+s8hyQbq92nmRAegjSEn5PqOJrr92XGlfbUxVtCQ1omyqrarDy5xys+vUY8mwVSIg6jGFnt8WIXvo6o+L3mCaLxYJdu3YhJSUFkZGR2L59Ozp37owdO3bgvPPOQ2mp57WEmpOWct6UiIj0w+kUHC+p/ON0n8ulZw4VlqOqxr/hx3HhJq/jp1LjwmEx+ZfkrN5xHPcuy4TNy9jdKEsIng3A2N2W8tnsd0/T2WefjXXr1qm9QXWWLVuGvn37BqxiLZGIoLS0FBEREfzmS6QBxhy1FgaDguRoC5KjLRh4RrzbbdU1Thw5UeGxmGdWfhmOnPA+IL2gzIGCMge2HCzyuK1ddJjHtfvS4sPRMdbqMSB99Y7juO2NLeoUcWe9/0sqqjH5jS1YfHN/DNXBpCe/k6bZs2fj5ptvxpEjR+B0OvHhhx/it99+w+uvv47PPvusKerYYogIysrKEB4ezjdwIg0w5kgPQowGpMSFIyUuHDjT/bbKqhocKix3S6jqfs8t8RyQDgBHiytxtLgSG/cVuJUbDQo6xljUhKpjjAXPrNrd4LI6QG25IsDMZZn4/qEhrf5UXaOWHPjyyy/xxBNPYOvWrXA6nejXrx8eeeQRDBs2rCnqGFBN2QXI6c9E2mLMETWs1F79x/pTLglVVn4ZiiuqAv5488eeg9F9OzTqvq329BwAZGRkICMjI9B1ISIiogCJMIegZ/to9Gwf7XFbUZkDWQW1Y6ey8svU3w8UlKHc4duVPVwZFODLX443OmlqKRqVNJF3iqLAYrHwNAGRRhhzRI0TE25CTLgJ/TrFuJWLCHJL7NifV4aHPvoZWV6WQfDGKcCJCv8umNwS+ZQ0xcTE+PymVFhYeFoVaskURUF0tGdGT0RNgzFHFFiKoiApKgxJUWE4MykSBwvK4MuVZAwK0MZiavoKBplPSdOCBQuauBqtg4jAZrMhKiqK33yJNMCYI2o6w85Owhe/HvNpW6cAGT05ew4AMGHChKauR6tQd/HQyMhIvoETaYAxR9R0RvRKxqMrfkVJRXWDs+eA2ovTR1lCMLxnslZVCxqfkiabzebzDpvzqHciIiLyTVioEc9e1weT39gCpYFlB5Tf/5l3XZ9Wv9wA4GPS1KZNm1N+ixMRKIqCmhr/R90TERFR8zOkRxIW39wfM5dloriBFcHnBWBF8JbCp6Tpm2++aep6tAqKonCRPSINMeaImt7QHkn4/qEh+O8vOfjyl2PIL6lEfGQYMnq2xfCevPZco2VmZqJPnz6B2l2TaCkLaBEREelFS/lsPu0ldIuLi7Fw4UL069cP6enpgahTiyUiKCwsRADzUCI6CcYckbb0HnONTpq+/vpr3HTTTUhOTsbzzz+PESNGYMuWLYGsW4sjInA4HLo9mIi0xpgj0pbeY86vFcEPHz6MpUuXYsmSJSgrK8P111+PqqoqLF++HD169GiqOhIREREFnc89TSNGjECPHj2wY8cOPP/88zh69Cief/75pqwbERERUbPhc0/TqlWrcM899+COO+5A165dm7JOLZaiKFyZmEhDjDkibek95nzuaVq3bh1KSkrQv39/nH/++fj3v/+NvLy8pqxbi6MoCqxWq24PJiKtMeaItKX3mPM5aRowYABeeukl5OTk4Pbbb8e7776L9u3bw+l0YvXq1SgpKWnKerYITqcT+fn5cDqdwa4KkS4w5oi0pfeY83v2nNVqxS233IL169fj559/xr333osnn3wSiYmJuOaaa5qiji1KdXV1sKtApCuMOSJt6TnmTmudpjPPPBNPPfUUDh8+jHfeeSdQdSIiIiJqdk57cUsAMBqNGDVqFD799NNA7I6IiIio2QlI0kS1FEVBTEyMbgfIEWmNMUekLb3HnF+LW9LJKYoCs9kc7GoQ6QZjjkhbeo859jQFkNPpxPHjx3U7q4BIa4w5Im3pPeaYNAWYXq/HQxQsjDkibek55pg0EREREfmASRMRERGRD5g0BZCiKIiLi9PtrAIirTHmiLSl95hj0hRAiqLAaDTq9mAi0hpjjkhbeo85Jk0B5HQ6kZubq9tZBURaY8wRaUvvMcekiYiIiMgHTJqIiIiIfMCkiYiIiMgHTJoCyGAwIDExEQYDn1YiLTDmiLSl95jTZ6ubiIigpqZG16ulEmmJMUekLb3HHJOmABIRFBQU6PZgItIaY45IW3qPOSZNRERERD5g0kRERETkAyZNAabXVVKJgoUxR6QtPcdcSLAr0JoYDAYkJSUFuxpEusGYI9KW3mOOPU0BJCKw2+26HSBHpDXGHJG29B5zTJoCSERQVFSk24OJSGuMOSJt6T3mmDQRERER+YBJExEREZEPmDQFWEgIx9YTaYkxR6QtPcecflveBAwGA+Lj44NdDSLdYMwRaUvvMceepgASEZSXl+t2gByR1hhzRNrSe8wxaQogEYHNZtPtwUSkNcYckbb0HnNMmoiIiIh8EPSkaeHChUhLS0NYWBjS09Oxbt26Brf98MMPMXToUCQkJCAqKgoDBgzAl19+qWFtiYiISK+CmjS99957mDZtGh5++GFs27YNgwYNwvDhw5Gdne11+7Vr12Lo0KFYuXIltm7diksvvRRXX301tm3bpnHNvVMUBSaTSdfX5SHSEmOOSFt6jzlFgnhi8vzzz0e/fv3w4osvqmXdu3fHqFGjMHfuXJ/2cfbZZ2Ps2LF45JFHfNreZrMhOjoaxcXFiIqKalS9iYiIKHBaymdz0JYccDgc2Lp1Kx588EG38mHDhmHjxo0+7cPpdKKkpASxsbENbmO322G329W/bTabel+n0wmgNnNWFAUi4ja47VTldfd3VVZWBqvV6paFN7S9wWDw2Le/5Y2tu69t8recbWKbtGyTiKC0tBQREREwGo2tok2NrTvbxDZp0SZFUVBSUoLw8HC3z7nTbZO3z9PmKGhJU35+PmpqajyulpyUlIRjx475tI958+ahrKwM119/fYPbzJ07F4899phHeV5eHiorKwEAFosF0dHRsNlsqKioULcJDw9HZGQkioqK4HA41PKoqChYrVYUFhaiurpaLY+OjkZZWRlKS0vdDqa4uDgYjUbk5ua61SExMRE1NTUoKChQyxRFQVJSEhwOB4qKitTykJAQxMfHo6KiQk38AMBkMiE2NhalpaUoKytTywPVppiYGJjNZuTl5bkd9GwT29Qc2uR0OlFcXIzo6Gi0bdu2VbSpNb5ObFPraVNCQgKKi4tRWloKg8EQsDaVlJSgJQja6bmjR4+iffv22LhxIwYMGKCWP/7443jjjTewa9euk97/nXfewa233opPPvkEQ4YMaXA7bz1NHTt2RFFRkdoFGKhMXkSQl5eH+Ph49WA62fZ6+nbCNrFNTdEmp9OJvLw8JCQkICQkpFW0qbF1Z5vYJi3aBADHjx9HQkKC2+fc6bbJZrMhJiaGp+caEh8fD6PR6NGrlJub69H7VN97772HSZMmYdmyZSdNmADAbDbDbDZ7lBsMBrcXHPjjhayvofL696876Lzt29v2jXnMpi73Vkd/y9kmtqkx5Y1tk6Iobr+3hjY1h7qzTWyTt307nU415k73M9S1vKF6NTdBq6XJZEJ6ejpWr17tVr569WoMHDiwwfu98847mDhxIt5++21ceeWVTV1NvyiKAovF4vXgIKLAY8wRaUvvMRfUa8/NmDEDN998M/r3748BAwZg8eLFyM7OxpQpUwAAs2bNwpEjR/D6668DqE2Yxo8fj+eeew4XXHCB2ktVd3402BRFaRb1INILxhyRtvQec0HtDxs7diwWLFiAOXPmoE+fPli7di1WrlyJlJQUAEBOTo7bmk3/+c9/UF1djbvuugvJycnqz9SpU4PVBDciguLiYq/ngIko8BhzRNrSe8wFdZ2mYGjKtSCcTidyc3ORmJjYYs7PErVkjDkibTVVzLWUdZr4LkNERETkAyZNRERERD5g0hRAiqJ4rJJKRE2HMUekLb3HXFBnz7U2iqIgMjIy2NUg0g3GHJG29B5z7GkKIBFBYWGhbmcVEGmNMUekLb3HHJOmABIROBwO3R5MRFpjzBFpS+8xx6SJiIiIyAdMmoiIiIh8wKQpgBRFQVRUlG5nFRBpjTFHpC29xxxnzwWQoiiwWq3BrgaRbjDmiLSl95hjT1MAOZ1O5Ofnw+l0BrsqRLrAmCPSlt5jjklTgFVXVwe7CkS6wpgj0paeY45JExEREZEPmDQRERER+YBJUwApioKYmBjdziog0hpjjkhbeo85zp4LIEVRYDabg10NIt1gzBFpS+8xx56mAHI6nTh+/LhuZxUQaY0xR6Qtvccck6YA0+v1eIiChTFHpC09xxyTJiIiIiIfMGkiIiIi8gGTpgBSFAVxcXG6nVVApDXGHJG29B5zTJoCSFEUGI1G3R5MRFpjzBFpS+8xx6QpgJxOJ3Jzc3U7q4BIa4w5Im3pPeaYNBERERH5gEkTERERkQ+YNBERERH5gElTABkMBiQmJsJg4NNKpAXGHJG29B5z+mx1ExER1NTU6Hq1VCItMeaItKX3mGPSFEAigoKCAt0eTERaY8wRaUvvMcekiYiIiMgHTJqIiIiIfMCkKcD0ukoqUbAw5oi0peeYCwl2BVoTg8GApKSkYFeDSDcYc0Ta0nvMsacpgEQEdrtdtwPkiLTGmCPSlt5jjklTAIkIioqKdHswEWmNMUekLb3HHJMmIiIiIh8waSIiIiLyAZOmAAsJ4dh6Ii0x5oi0peeY02/Lm4DBYEB8fHywq0GkG4w5Im3pPebY0xRAIoLy8nLdDpAj0hpjjkhbeo85Jk0BJCKw2Wy6PZiItMaYI9KW3mOOSRMRERGRD5g0EREREfmASVMAKYoCk8mk6+vyEGmJMUekLb3HHGfPBZCiKIiNjQ12NYh0gzFHpC29xxx7mgJIRFBSUqLbAXJEWmPMEWlL7zHHpCmARARlZWW6PZiItMaYI9KW3mOOSRMRERGRD5g0EREREfmASVMAKYoCi8Wi21kFRFpjzBFpS+8xF/SkaeHChUhLS0NYWBjS09Oxbt26k27/7bffIj09HWFhYejcuTMWLVqkUU1PTVEUREdH6/ZgItIaY45IW3qPuaAmTe+99x6mTZuGhx9+GNu2bcOgQYMwfPhwZGdne90+KysLI0aMwKBBg7Bt2zY89NBDuOeee7B8+XKNa+6diKC4uFi3A+SItMaYI9KW3mNOkSC2/Pzzz0e/fv3w4osvqmXdu3fHqFGjMHfuXI/tH3jgAXz66afYuXOnWjZlyhRs374dmzZt8ukxbTYboqOjUVxcjKioqNNvhAun04nc3FwkJibCYAh6Jx5Rq8eYI9JWU8VcU342B1LQ3mUcDge2bt2KYcOGuZUPGzYMGzdu9HqfTZs2eWyfkZGBLVu2oKqqqsnqSkRERBS0FcHz8/NRU1ODpKQkt/KkpCQcO3bM632OHTvmdfvq6mrk5+cjOTnZ4z52ux12u139u7i4GABw4sQJOJ1OALXnaBVFgYi4dTmeqrzu/nXqFv0ymUxuGXhD2xsMBo99+1ve2Lr72iZ/y9kmtknLNjmdTthsNphMJoSEhLSKNjW27mwT26RFmwCoMef6OXe6bbLZbADQ7E/7Bf0yKvUHk4nISQeYedveW3mduXPn4rHHHvMoT0lJ8beqRERE1IRKSkoQHR0d7Go0KGhJU3x8PIxGo0evUm5urkdvUp22bdt63T4kJARxcXFe7zNr1izMmDFD/dvpdKKwsBBxcXEBH/1vs9nQsWNHHDp0qFmfkyVqLRhzRNpqqpirO1PTrl27gO2zKQQtaTKZTEhPT8fq1asxevRotXz16tUYOXKk1/sMGDAAK1ascCtbtWoV+vfvj9DQUK/3MZvNMJvNbmVt2rQ5vcqfQlRUFN/AiTTEmCPSVlPEXHPuYaoT1OkmM2bMwMsvv4wlS5Zg586dmD59OrKzszFlyhQAtb1E48ePV7efMmUKDh48iBkzZmDnzp1YsmQJXnnlFcycOTNYTSAiIiKdCOqYprFjx6KgoABz5sxBTk4OevbsiZUrV6rjjXJyctzWbEpLS8PKlSsxffp0vPDCC2jXrh3+9a9/YcyYMcFqAhEREelEUNdpam3sdjvmzp2LWbNmeZwSJKLAY8wRaUvvMcekiYiIiMgHXEKXiIiIyAdMmoiIiIh8wKSJSAcuueQSTJs27bT2sXTp0iZfroOIqDnTfdKUmpqKBQsWaPZ4gfjwCqRHH30Uffr0CXY1qAUYO3Ysdu/ereljMj4Zn9SwNWvWQFEUnDhxwuf7aH2MT5w4EaNGjdLs8U7ldL/8Bf0yKqfD4XDAZDI1+ePU1NRAURReRZ10q6qqChaLBRaLJdhV8cD4JAosEUFNTQ1CQlp0itA0pBkZPHiw3HXXXXLXXXdJdHS0xMbGysMPPyxOp1NERFJSUuTvf/+7TJgwQaKiomT8+PEiIrJhwwYZNGiQhIWFSYcOHeSvf/2rlJaW+vR4ANx+REReffVViY6OlhUrVkj37t3FaDTK/v37xW63y3333Sft2rUTq9Uq5513nnzzzTfq/vLz82XcuHHSvn17sVgs0rNnT3n77bfV2ydMmODxeFlZWfLNN98IAPniiy+kT58+EhYWJpdeeqkcP35cVq5cKWeddZZERkbKuHHjpKysTN2f0+mUf/7zn5KWliZhYWHSu3dvWbZsmXp73X6/+uorSU9PF4vFIgMGDJBdu3ap7axfn1dffbXRrx81D6WlpXLzzTdLeHi4tG3bVp555hkZPHiwTJ06VUREAMhHH33kdp/o6Gj1tc/KyhIA8t5778ngwYPFbDbLkiVL1LioM3v2bDnnnHPk9ddfl5SUFImKipKxY8eKzWZTt7HZbHLDDTeI1WqVtm3byrPPPutWl5NhfDI+W4NTfa4VFhbKzTffLG3atBGLxSJXXHGF7N69W73/gQMH5KqrrpI2bdqI1WqVHj16yOeff67GqevPhAkTTloXX47x9PR0CQ0Nla+//vqUx3B1dbXccsstkpqaKmFhYdKtWzdZsGCBevvs2bM9Hu+bb75xe4+56KKLJCwsTPr37y+//fabbN68WdLT0yU8PFwyMjIkNzfXrQ1LliyRs846S8xms5x55pnywgsvqLfV7Xf58uVyySWXiMVikd69e8vGjRtF5I+Yc/2ZPXu2X69ns0uaIiIiZOrUqbJr1y558803xWq1yuLFi0VE1Dfmp59+Wvbs2SN79uyRn376SSIiImT+/Pmye/du2bBhg/Tt21cmTpx4yscrKCiQDh06yJw5cyQnJ0dycnJEpPbNKjQ0VAYOHCgbNmyQXbt2SWlpqdxwww0ycOBAWbt2rezdu1eefvppMZvN6gF++PBhefrpp2Xbtm2yb98++de//iVGo1G+++47ERE5ceKEDBgwQCZPnqw+XnV1tfpCXnDBBbJ+/Xr58ccfpUuXLjJ48GAZNmyY/Pjjj7J27VqJi4uTJ598Uq3/Qw89JGeddZZ88cUXsm/fPnn11VfFbDbLmjVrROSPA+T888+XNWvWyK+//iqDBg2SgQMHiohIeXm53HvvvXL22Wer9SkvLw/cC0pBcccdd0iHDh1k1apV8tNPP8lVV12lxpWI70lTamqqLF++XPbv3y9HjhzxmjRFRETIn/70J/n5559l7dq10rZtW3nooYfUbW699VZJSUmRr776Sn7++WcZPXq0REZG+pQ0MT4Zn63BqT7XrrnmGunevbusXbtWMjMzJSMjQ7p06SIOh0NERK688koZOnSo/PTTT7Jv3z5ZsWKFfPvtt1JdXS3Lly8XAPLbb79JTk6OnDhx4qR1OdUx3rt3b1m1apXs3btX8vPzT3kMOxwOeeSRR2Tz5s2yf/9+tW3vvfeeiIiUlJTI9ddfL1dccYX6eHa7XX2Pqdv3jh075IILLpB+/frJJZdc4hZnU6ZMUeu/ePFiSU5OVt+Xli9fLrGxsbJ06VIREbf9fvbZZ/Lbb7/JtddeKykpKVJVVSV2u10WLFggUVFRan1KSkr8ej2bXdLUvXt3NQMXEXnggQeke/fuIlKbNI0aNcrtPjfffLPcdtttbmXr1q0Tg8EgFRUVp3zMlJQUmT9/vltZ3Te8zMxMtWzv3r2iKIocOXLEbdvLL79cZs2a1eD+R4wYIffee69bG+t/YLh+46wzd+5cASD79u1Ty26//XbJyMgQkdrehLCwMDWDrjNp0iT585//3OB+P//8cwGgPjd1vQXUOpSUlIjJZJJ3331XLSsoKBCLxeJ30uT6jVFEvCZNVqvVrWfpvvvuk/PPP19EanuZQkND3b6ZnjhxQqxWq09Jkwjjk/HZ8p3sc2337t0CQDZs2KDelp+fLxaLRd5//30REenVq5c8+uijXvdddwwVFRX5VZ+GjvGPP/5YLfPlGPbmzjvvlDFjxqh/T5gwQUaOHOm2Td17zMsvv6yWvfPOOwJA/ve//6llc+fOlTPPPFP9u2PHjm69wyIif//732XAgAEN7vfXX38VALJz504R8Xwf81ezO2F5wQUXQFEU9e8BAwZg3rx5qKmpAQD079/fbfutW7di7969eOutt9QyEYHT6URWVha6d+/eqHqYTCb07t1b/fvHH3+EiKBbt25u29ntdsTFxQGoHVvx5JNP4r333sORI0dgt9tht9sRHh7u02O6Pl5SUhKsVis6d+7sVrZ582YAwI4dO1BZWYmhQ4e67cPhcKBv374N7jc5ORkAkJubi06dOvlUL2o59u3bB4fDgQEDBqhlsbGxOPPMM/3eV/1Y8yY1NRWRkZHq38nJycjNzQUA7N+/H1VVVTjvvPPU26OjoxtVl/oYn9SSNPS5tmPHDoSEhOD8889Xb4uLi8OZZ56JnTt3AgDuuece3HHHHVi1ahWGDBmCMWPGuB0zgeQa874ew4sWLcLLL7+MgwcPoqKiAg6Hw+fJC/VjCgB69erlVlb3fpKXl4dDhw5h0qRJmDx5srpNdXW1x4V+G4qps846y6d6nUyzS5pOpf4bnNPpxO2334577rnHY9vTedOxWCxuB7nT6YTRaMTWrVthNBrdto2IiAAAzJs3D/Pnz8eCBQvQq1cvhIeHY9q0aXA4HD49ZmhoqPq7oihuf9eVOZ1OtT4A8Pnnn6N9+/Zu29Vf2r7+fl3vT62L+LDAv6IoHttVVVV5bOdLMnGyY7TuMVzjyNc6ngrjk1ozEVGPhVtvvRUZGRn4/PPPsWrVKsydOxfz5s3DX//614A/rmvM+3IMv//++5g+fTrmzZuHAQMGIDIyEk8//TS+//57nx7P27Ffv6x+TL300ktuSSYAj5hvyphqdknTd9995/F3165dPZ6UOv369cOvv/6KLl26NOrxTCaT2ot1Mn379kVNTQ1yc3MxaNAgr9usW7cOI0eOxE033QSg9kXas2ePW2+Xr493Kj169IDZbEZ2djYGDx7c6P0Eqj7UPHTp0gWhoaH47rvv1C8NRUVF2L17t3qcJCQkICcnR73Pnj17UF5eHvC6nHHGGQgNDcXmzZvRsWNHAIDNZsOePXt8PmYZn4zP1qChz7UePXqguroa33//PQYOHAgAKCgowO7du92Oy44dO2LKlCmYMmUKZs2ahZdeegl//etf1dnj/hwjvh5TvhzD69atw8CBA3HnnXeqZfv27WvU451KUlIS2rdvj/379+PGG29s9H5Otz7NLmk6dOgQZsyYgdtvvx0//vgjnn/+ecybN6/B7R944AFccMEFuOuuuzB58mSEh4dj586dWL16NZ5//vlTPl5qairWrl2LcePGwWw2Iz4+3ut23bp1w4033ojx48dj3rx56Nu3L/Lz8/H111+jV69eGDFiBLp06YLly5dj48aNiImJwbPPPotjx465Hfypqan4/vvvceDAAURERCA2Ntb/JwlAZGQkZs6cienTp8PpdOKiiy6CzWbDxo0bERERgQkTJvi0n9TUVGRlZSEzMxMdOnRAZGSkLi/C2FpERERg0qRJuO+++xAXF4ekpCQ8/PDDbtPxL7vsMvz73//GBRdcAKfTiQceeMCj1yQQIiMjMWHCBNx3332IjY1FYmIiZs+eDYPB4NH71BDGJ+OzNWjoc61r164YOXIkJk+ejP/85z+IjIzEgw8+iPbt22PkyJEAgGnTpmH48OHo1q0bioqK8PXXX6vHbEpKChRFwWeffYYRI0bAYrGoPasN8fUY9+UY7tKlC15//XV8+eWXSEtLwxtvvIEffvgBaWlpbo/35Zdf4rfffkNcXJzHqTR/PProo7jnnnsQFRWF4cOHw263Y8uWLSgqKsKMGTN82kdqaipKS0vxv//9D+eccw6sViusVqvvlWj0aKgmMHjwYLnzzjtlypQpEhUVJTExMfLggw+6LTlQf1CoiMjmzZtl6NChEhERIeHh4dK7d295/PHHfXrMTZs2Se/evcVsNntMaa6vbqZAamqqhIaGStu2bWX06NHy008/iUjtgNuRI0dKRESEJCYmyv/93//J+PHj3QbB/fbbb3LBBReIxWLxmO7pOpjPWx3qDwp1Op3y3HPPyZlnnimhoaGSkJAgGRkZ8u2334qI90GC27ZtUx9XRKSyslLGjBkjbdq04ZTmVqKkpERuuukmsVqtkpSUJE899ZTb4M8jR47IsGHDJDw8XLp27SorV670OhB827ZtbvttaMkBV/Pnz5eUlBT1b29LDpx33nny4IMP+tQWxifjs6U71eda3ZID0dHRYrFYJCMjw23JgbvvvlvOOOMMMZvNkpCQIDfffLPk5+ert8+ZM0fatm0riqKccskBEd+PcZFTH8OVlZUyceJEiY6OljZt2sgdd9whDz74oFsc5Obmqp/PqLfkgOt7jK9x9tZbb0mfPn3EZDJJTEyMXHzxxfLhhx+KiPf3rqKiIvVx60yZMkXi4uIateSAIhKAAQYBcskll6BPnz6argBMRNopKytD+/btMW/ePEyaNCnY1SFqcvxca12a3ek5Imo9tm3bhl27duG8885DcXEx5syZAwDqqQciopakVSdN69atw/Dhwxu8vbS0VMPaEOnTM888g99++w0mkwnp6elYt24d4uPjGZ9EAZadnY0ePXo0ePuOHTu4lMVpalan5wKtoqICR44cafD2xs64I6LTx/gkCqzq6mocOHCgwdtTU1N5PbnT1KqTJiIiIqJA4WXBiYiIiHzApImIiIjIB0yaiIiIiHzApImIiIjIB0yaiIiIiHzApImINJebm4vbb78dnTp1gtlsRtu2bZGRkYFNmzYBqL0y+ccffxzcShIR1cMFG4hIc2PGjEFVVRVee+01dO7cGcePH8f//vc/FBYWBrtqREQNYk8TEWnqxIkTWL9+Pf75z3/i0ksvRUpKCs477zzMmjULV155JVJTUwEAo0ePhqIo6t/79u3DyJEjkZSUhIiICJx77rn46quv3Padk5ODK6+8EhaLBWlpaXj77beRmprqdt2v4uJi3HbbbUhMTERUVBQuu+wybN++XaPWE1FLxqSJiDQVERGBiIgIfPzxx7Db7R63//DDDwCAV199FTk5OerfpaWlGDFiBL766its27YNGRkZuPrqq5Gdna3ed/z48Th69CjWrFmD5cuXY/HixcjNzVVvFxFceeWVOHbsGFauXImtW7eiX79+uPzyy9nLRUSnxBXBiUhzy5cvx+TJk1FRUYF+/fph8ODBGDduHHr37g2gdkzTRx99hFGjRp10P2effTbuuOMO3H333di1axe6d++OH374Af379wcA7N27F127dsX8+fMxbdo0fP311xg9ejRyc3NhNpvV/XTp0gX3338/brvttiZrMxG1fOxpIiLNjRkzBkePHsWnn36KjIwMrFmzBv369cPSpUsbvE9ZWRnuv/9+9OjRA23atEFERAR27dql9jT99ttvCAkJQb9+/dT7dOnSBTExMerfW7duRWlpKeLi4tQer4iICGRlZWHfvn1N1l4iah04EJyIgiIsLAxDhw7F0KFD8cgjj+DWW2/F7NmzMXHiRK/b33ffffjyyy/xzDPPoEuXLrBYLLj22mvhcDgA1J5688a13Ol0Ijk5GWvWrPHYrk2bNqfbJCJq5Zg0EVGz0KNHD3WZgdDQUNTU1Ljdvm7dOkycOBGjR48GUDvGyfWK7meddRaqq6uxbds2pKenA6g9PXfixAl1m379+uHYsWMICQlRB5gTEfmKp+eISFMFBQW47LLL8Oabb+Knn35CVlYWli1bhqeeegojR44EAKSmpuJ///sfjh07hqKiIgC1p9o+/PBDZGZmYvv27bjhhhvgdDrV/Z511lkYMmQIbrvtNmzevBnbtm3DbbfdBovFAkVRAABDhgzBgAEDMGrUKHz55Zc4cOAANm7ciP/7v//Dli1btH8yiKhFYdJERJqKiIjA+eefj/nz5+Piiy9Gz5498be//Q2TJ0/Gv//9bwDAvHnzsHr1anTs2BF9+/YFAMyfPx8xMTEYOHAgrr76amRkZLiNXwKA119/HUlJSbj44osxevRoTJ48GZGRkQgLCwNQO8B85cqVuPjii3HLLbegW7duGDduHA4cOICkpCRtnwgianE4e46IWq3Dhw+jY8eO+Oqrr3D55ZcHuzpE1MIxaSKiVuPrr79GaWkpevXqhZycHNx///04cuQIdu/ejdDQ0GBXj4haOA4EJ6JWo6qqCg899BD279+PyMhIDBw4EG+99RYTJiIKCPY0EREREfmAA8GJiIiIfMCkiYiIiMgHTJqIiIiIfMCkiYiIiMgHTJqIiIiIfMCkiYiIiMgHTJqIiIiIfMCkiYiIiMgHTJqIiIiIfPD/hoJa8BRuANEAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 600x400 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot allele frequency changes over timepoints\n",
    "timepoint_order = [\"pre_treatment\", \"during_treatment\", \"post_treatment\"]\n",
    "selected_variant[\"sample_label\"] = pd.Categorical(selected_variant[\"sample_label\"],categories=timepoint_order, ordered=True)\n",
    "selected_variant_sorted = selected_variant.sort_values(\"sample_label\")\n",
    "plt.figure(figsize=(6, 4))\n",
    "plt.plot(selected_variant_sorted[\"sample_label\"],selected_variant_sorted[\"allele_frequency\"],marker=\"o\",linewidth=2, markersize=8,)\n",
    "plt.title(\"KRAS Variant Allele Frequency Over Stages\", fontsize=12, weight=\"bold\")\n",
    "plt.xlabel(\"Stage\", fontsize=10)\n",
    "plt.ylabel(\"Allele Frequency\", fontsize=10)\n",
    "plt.ylim(0, 1)\n",
    "plt.grid(True, alpha=0.3, linestyle=\"--\")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "872dadc1",
   "metadata": {},
   "source": [
    "### Analysis Summary:\n",
    "_____\n",
    "\n",
    "\n",
    "**KRAS variant at position 25,209,843 (chr12)** was identified (read depth â‰¥20x and allele frequency â‰¥10%) as a persistent mutation across all treatment stages in patient CTC030. \n",
    "\n",
    "**KRAS variant at position 25,209,843 (chr12)** has potential as a longitudinal biomarker for colorectal cancer monitoring.\n",
    "\n",
    "**Limitations:** These findings are based on a single patient case study. This pipeline is explicitly designed to identify variants in circulating tumor DNA for single patients and for a given gene. For this reason, no conclusions can be extrapolated to a population level."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d173094",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "vca_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
